{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "120ef59f-945a-4fe5-9337-994e062657ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import csv\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "c9f8578c-6ee9-480e-bd17-28f57b3cbc99",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "###\n",
    "#LEGACY\n",
    "###\n",
    "\n",
    "# with open('./data/high_resolution_2_membrane_protein_sequences_after_2000.csv', newline='') as f:\n",
    "#     reader = csv.reader(f)\n",
    "#     file = list(reader)  # convert the reader to a list of rows\n",
    "# membrane_pdb_ids = []\n",
    "# for line in file[2:]:\n",
    "#     if len(line[0])==4:\n",
    "#         pdb_id = line[0]\n",
    "#         membrane_pdb_ids.append(pdb_id)\n",
    "#     elif len(line[0])>1:\n",
    "#         pdb_id = line[0][:2]+line[0][-2:]\n",
    "#         membrane_pdb_ids.append(pdb_id)\n",
    "        \n",
    "# with open('./data/high_resolution_2_protein_sequences_after_2000.csv', newline='') as f:\n",
    "#     reader = csv.reader(f)\n",
    "#     file = list(reader)  # convert the reader to a list of rows\n",
    "\n",
    "# fasta = []\n",
    "# names = []\n",
    "# for line in file[2:]:\n",
    "#     if len(line[0])==4:\n",
    "#         pdb_id = line[0]\n",
    "#     elif len(line[0])>1:\n",
    "#         pdb_id = line[0][:2]+line[0][-2:]\n",
    "#     if not len(line[2]) or line[0] == 'Entry ID' or line[-2] != 'polypeptide(L)' or pdb_id in membrane_pdb_ids:\n",
    "#         continue\n",
    "#     #line = line.split(',')\n",
    "#     chains = [i.strip() for i in line[2].split(',')]\n",
    "#     if len(chains)-1:\n",
    "#         for c in chains:\n",
    "#             fasta.append(f'>{pdb_id}_{c}')\n",
    "#             names.append(f'>{pdb_id}_{c}')\n",
    "#             fasta.append(line[6])\n",
    "#     else:\n",
    "#         fasta.append(f'>{pdb_id}_{chains[0]}')\n",
    "#         names.append(f'>{pdb_id}_{chains[0]}')\n",
    "#         fasta.append(line[6])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "38af7f81-d4ab-45d8-8cb5-6dba3e7f478e",
   "metadata": {},
   "outputs": [],
   "source": [
    "path_mem_proteins = 'data/high_resolution_2dot8_membrane_protein_sequences_before_2000.csv'\n",
    "with open(path_mem_proteins, newline='') as f:\n",
    "    reader = csv.reader(f)\n",
    "    mem_file = list(reader)  # convert the reader to a list of rows\n",
    "membrane_pdb_ids = np.array(mem_file)[:,0]\n",
    "\n",
    "path_all_proteins = './data/high_resolution_2dot8_protein_sequences_before_2000.csv'\n",
    "with open(path_all_proteins, newline='') as f:\n",
    "    reader = csv.reader(f)\n",
    "    prot_file = list(reader)  # convert the reader to a list of rows"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "846cfcd3-fe11-41ba-9948-0a4d533746f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "file = prot_file\n",
    "fasta = []\n",
    "names = []\n",
    "for line in file:\n",
    "    pdb_id = line[0]\n",
    "    if line[-1] != 'polypeptide(L)' or pdb_id in membrane_pdb_ids:\n",
    "        continue\n",
    "    #line = line.split(',')\n",
    "    chains = [i.strip() for i in line[1].split(',')]\n",
    "    if len(chains)-1:\n",
    "        for c in chains:\n",
    "            fasta.append(f'>{pdb_id}_{c}')\n",
    "            names.append(f'>{pdb_id}_{c}')\n",
    "            fasta.append(line[2])\n",
    "    else:\n",
    "        fasta.append(f'>{pdb_id}_{chains[0]}')\n",
    "        names.append(f'>{pdb_id}_{chains[0]}')\n",
    "        fasta.append(line[2])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "20fbfb0e-4238-437f-977c-9d453585c9eb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "13718 13718\n"
     ]
    }
   ],
   "source": [
    "un, ind, counts = np.unique(names, return_counts=True, return_index=True)\n",
    "print(len(names), len(un))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "63429e97-30da-4ddb-bc5f-2b09366719b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('./fasta/high_resolution_2dot8_protein_sequences_before_2000.fasta', 'w') as f:\n",
    "    for line in fasta:\n",
    "        f.write(f'{line}\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "6983d69f-279e-4669-9e49-f81978f97da9",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "createdb ./fasta/high_resolution_2dot8_protein_sequences_before_2000.fasta ./mmseq_db/inputDB \n",
      "\n",
      "MMseqs Version:       \t13.45111\n",
      "Database type         \t0\n",
      "Shuffle input database\ttrue\n",
      "Createdb mode         \t0\n",
      "Write lookup file     \t1\n",
      "Offset of numeric ids \t0\n",
      "Compressed            \t0\n",
      "Verbosity             \t3\n",
      "\n",
      "Converting sequences\n",
      "[=\n",
      "Time for merging to inputDB_h: 0h 0m 0s 1ms\n",
      "Time for merging to inputDB: 0h 0m 0s 2ms\n",
      "Database type: Aminoacid\n",
      "Time for processing: 0h 0m 0s 140ms\n",
      "cluster ./mmseq_db/inputDB ./mmseq_db/clusterDB ./mmseq_db/tmpDir --min-seq-id 0.3 -c 0.8 --cov-mode 1 \n",
      "\n",
      "MMseqs Version:                     \t13.45111\n",
      "Substitution matrix                 \tnucl:nucleotide.out,aa:blosum62.out\n",
      "Seed substitution matrix            \tnucl:nucleotide.out,aa:VTML80.out\n",
      "Sensitivity                         \t4\n",
      "k-mer length                        \t0\n",
      "k-score                             \t2147483647\n",
      "Alphabet size                       \tnucl:5,aa:21\n",
      "Max sequence length                 \t65535\n",
      "Max results per query               \t20\n",
      "Split database                      \t0\n",
      "Split mode                          \t2\n",
      "Split memory limit                  \t0\n",
      "Coverage threshold                  \t0.8\n",
      "Coverage mode                       \t1\n",
      "Compositional bias                  \t1\n",
      "Diagonal scoring                    \ttrue\n",
      "Exact k-mer matching                \t0\n",
      "Mask residues                       \t1\n",
      "Mask lower case residues            \t0\n",
      "Minimum diagonal score              \t15\n",
      "Include identical seq. id.          \tfalse\n",
      "Spaced k-mers                       \t1\n",
      "Preload mode                        \t0\n",
      "Pseudo count a                      \t1\n",
      "Pseudo count b                      \t1.5\n",
      "Spaced k-mer pattern                \t\n",
      "Local temporary path                \t\n",
      "Threads                             \t22\n",
      "Compressed                          \t0\n",
      "Verbosity                           \t3\n",
      "Add backtrace                       \tfalse\n",
      "Alignment mode                      \t3\n",
      "Alignment mode                      \t0\n",
      "Allow wrapped scoring               \tfalse\n",
      "E-value threshold                   \t0.001\n",
      "Seq. id. threshold                  \t0.3\n",
      "Min alignment length                \t0\n",
      "Seq. id. mode                       \t0\n",
      "Alternative alignments              \t0\n",
      "Max reject                          \t2147483647\n",
      "Max accept                          \t2147483647\n",
      "Score bias                          \t0\n",
      "Realign hits                        \tfalse\n",
      "Realign score bias                  \t-0.2\n",
      "Realign max seqs                    \t2147483647\n",
      "Gap open cost                       \tnucl:5,aa:11\n",
      "Gap extension cost                  \tnucl:2,aa:1\n",
      "Zdrop                               \t40\n",
      "Rescore mode                        \t0\n",
      "Remove hits by seq. id. and coverage\tfalse\n",
      "Sort results                        \t0\n",
      "Cluster mode                        \t0\n",
      "Max connected component depth       \t1000\n",
      "Similarity type                     \t2\n",
      "Single step clustering              \tfalse\n",
      "Cascaded clustering steps           \t3\n",
      "Cluster reassign                    \tfalse\n",
      "Remove temporary files              \tfalse\n",
      "Force restart with latest tmp       \tfalse\n",
      "MPI runner                          \t\n",
      "k-mers per sequence                 \t21\n",
      "Scale k-mers per sequence           \tnucl:0.200,aa:0.000\n",
      "Adjust k-mer length                 \tfalse\n",
      "Shift hash                          \t67\n",
      "Include only extendable             \tfalse\n",
      "Skip repeating k-mers               \tfalse\n",
      "\n",
      "Set cluster sensitivity to -s 5.000000\n",
      "Set cluster mode GREEDY MEM\n",
      "Set cluster iterations to 3\n",
      "linclust ./mmseq_db/inputDB ./mmseq_db/tmpDir/17033438874298005812/clu_redundancy ./mmseq_db/tmpDir/17033438874298005812/linclust --cluster-mode 3 --max-iterations 1000 --similarity-type 2 --threads 22 --compressed 0 -v 3 --sub-mat nucl:nucleotide.out,aa:blosum62.out -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.3 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 1 --max-seq-len 65535 --comp-bias-corr 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca 1 --pcb 1.5 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --gap-open nucl:5,aa:11 --gap-extend nucl:2,aa:1 --zdrop 40 --alph-size nucl:5,aa:13 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale nucl:0.200,aa:0.000 --adjust-kmer-len 0 --mask 0 --mask-lower-case 0 -k 0 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --rescore-mode 0 --filter-hits 0 --sort-results 0 --remove-tmp-files 0 --force-reuse 0 \n",
      "\n",
      "kmermatcher ./mmseq_db/inputDB ./mmseq_db/tmpDir/17033438874298005812/linclust/389093368727786622/pref --sub-mat nucl:nucleotide.out,aa:blosum62.out --alph-size nucl:5,aa:13 --min-seq-id 0.3 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale nucl:0.200,aa:0.000 --adjust-kmer-len 0 --mask 0 --mask-lower-case 0 --cov-mode 1 -k 0 -c 0.8 --max-seq-len 65535 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --threads 22 --compressed 0 -v 3 \n",
      "\n",
      "kmermatcher ./mmseq_db/inputDB ./mmseq_db/tmpDir/17033438874298005812/linclust/389093368727786622/pref --sub-mat nucl:nucleotide.out,aa:blosum62.out --alph-size nucl:5,aa:13 --min-seq-id 0.3 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale nucl:0.200,aa:0.000 --adjust-kmer-len 0 --mask 0 --mask-lower-case 0 --cov-mode 1 -k 0 -c 0.8 --max-seq-len 65535 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --threads 22 --compressed 0 -v 3 \n",
      "\n",
      "Database size: 13718 type: Aminoacid\n",
      "Reduced amino acid alphabet: (A S T) (C) (D B N) (E Q Z) (F Y) (G) (H) (I V) (K R) (L J M) (P) (W) (X) \n",
      "\n",
      "Generate k-mers list for 1 split\n",
      "[=================================================================] 13.72K 0s 17ms\n",
      "Sort kmer 0h 0m 0s 47ms\n",
      "Sort by rep. sequence 0h 0m 0s 28ms\n",
      "Time for fill: 0h 0m 0s 5ms\n",
      "Time for merging to pref: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 146ms\n",
      "rescorediagonal ./mmseq_db/inputDB ./mmseq_db/inputDB ./mmseq_db/tmpDir/17033438874298005812/linclust/389093368727786622/pref ./mmseq_db/tmpDir/17033438874298005812/linclust/389093368727786622/pref_rescore1 --sub-mat nucl:nucleotide.out,aa:blosum62.out --rescore-mode 0 --wrapped-scoring 0 --filter-hits 0 -e 0.001 -c 0.8 -a 0 --cov-mode 1 --min-seq-id 0.5 --min-aln-len 0 --seq-id-mode 0 --add-self-matches 0 --sort-results 0 --db-load-mode 0 --threads 22 --compressed 0 -v 3 \n",
      "\n",
      "[================================================================] 13.72K 0s 8ms\n",
      "=Time for merging to pref_rescore1: 0h 0m 0s 10ms\n",
      "Time for processing: 0h 0m 0s 74ms\n",
      "clust ./mmseq_db/inputDB ./mmseq_db/tmpDir/17033438874298005812/linclust/389093368727786622/pref_rescore1 ./mmseq_db/tmpDir/17033438874298005812/linclust/389093368727786622/pre_clust --cluster-mode 3 --max-iterations 1000 --similarity-type 2 --threads 22 --compressed 0 -v 3 \n",
      "\n",
      "Clustering mode: Greedy Low Mem\n",
      "Total time: 0h 0m 0s 26ms\n",
      "\n",
      "Size of the sequence database: 13718\n",
      "Size of the alignment database: 13718\n",
      "Number of clusters: 2145\n",
      "\n",
      "Writing results 0h 0m 0s 0ms\n",
      "Time for merging to pre_clust: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 30ms\n",
      "createsubdb ./mmseq_db/tmpDir/17033438874298005812/linclust/389093368727786622/order_redundancy ./mmseq_db/inputDB ./mmseq_db/tmpDir/17033438874298005812/linclust/389093368727786622/input_step_redundancy -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step_redundancy: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 1ms\n",
      "createsubdb ./mmseq_db/tmpDir/17033438874298005812/linclust/389093368727786622/order_redundancy ./mmseq_db/tmpDir/17033438874298005812/linclust/389093368727786622/pref ./mmseq_db/tmpDir/17033438874298005812/linclust/389093368727786622/pref_filter1 -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to pref_filter1: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 2ms\n",
      "filterdb ./mmseq_db/tmpDir/17033438874298005812/linclust/389093368727786622/pref_filter1 ./mmseq_db/tmpDir/17033438874298005812/linclust/389093368727786622/pref_filter2 --filter-file ./mmseq_db/tmpDir/17033438874298005812/linclust/389093368727786622/order_redundancy --threads 22 --compressed 0 -v 3 \n",
      "\n",
      "Filtering using file(s)\n",
      "[=================================================================] 2.14K 0s 3ms\n",
      "Time for merging to pref_filter2: 0h 0m 0s 6ms\n",
      "Time for processing: 0h 0m 0s 20ms\n",
      "rescorediagonal ./mmseq_db/tmpDir/17033438874298005812/linclust/389093368727786622/input_step_redundancy ./mmseq_db/tmpDir/17033438874298005812/linclust/389093368727786622/input_step_redundancy ./mmseq_db/tmpDir/17033438874298005812/linclust/389093368727786622/pref_filter2 ./mmseq_db/tmpDir/17033438874298005812/linclust/389093368727786622/pref_rescore2 --sub-mat nucl:nucleotide.out,aa:blosum62.out --rescore-mode 1 --wrapped-scoring 0 --filter-hits 1 -e 0.001 -c 0.8 -a 0 --cov-mode 1 --min-seq-id 0.3 --min-aln-len 0 --seq-id-mode 0 --add-self-matches 0 --sort-results 0 --db-load-mode 0 --threads 22 --compressed 0 -v 3 \n",
      "\n",
      "[=================================================================] 2.14K 0s 7ms\n",
      "Time for merging to pref_rescore2: 0h 0m 0s 1ms\n",
      "Time for processing: 0h 0m 0s 43ms\n",
      "align ./mmseq_db/tmpDir/17033438874298005812/linclust/389093368727786622/input_step_redundancy ./mmseq_db/tmpDir/17033438874298005812/linclust/389093368727786622/input_step_redundancy ./mmseq_db/tmpDir/17033438874298005812/linclust/389093368727786622/pref_rescore2 ./mmseq_db/tmpDir/17033438874298005812/linclust/389093368727786622/aln --sub-mat nucl:nucleotide.out,aa:blosum62.out -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.3 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 1 --max-seq-len 65535 --comp-bias-corr 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca 1 --pcb 1.5 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --gap-open nucl:5,aa:11 --gap-extend nucl:2,aa:1 --zdrop 40 --threads 22 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 2145 type: Aminoacid\n",
      "Target database size: 2145 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 2.14K 0s 17ms\n",
      "Time for merging to aln: 0h 0m 0s 1ms\n",
      "2373 alignments calculated\n",
      "2344 sequence pairs passed the thresholds (0.987779 of overall calculated)\n",
      "1.092774 hits per query sequence\n",
      "Time for processing: 0h 0m 0s 26ms\n",
      "clust ./mmseq_db/tmpDir/17033438874298005812/linclust/389093368727786622/input_step_redundancy ./mmseq_db/tmpDir/17033438874298005812/linclust/389093368727786622/aln ./mmseq_db/tmpDir/17033438874298005812/linclust/389093368727786622/clust --cluster-mode 3 --max-iterations 1000 --similarity-type 2 --threads 22 --compressed 0 -v 3 \n",
      "\n",
      "Clustering mode: Greedy Low Mem\n",
      "Total time: 0h 0m 0s 21ms\n",
      "\n",
      "Size of the sequence database: 2145\n",
      "Size of the alignment database: 2145\n",
      "Number of clusters: 1982\n",
      "\n",
      "Writing results 0h 0m 0s 0ms\n",
      "Time for merging to clust: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 25ms\n",
      "mergeclusters ./mmseq_db/inputDB ./mmseq_db/tmpDir/17033438874298005812/clu_redundancy ./mmseq_db/tmpDir/17033438874298005812/linclust/389093368727786622/pre_clust ./mmseq_db/tmpDir/17033438874298005812/linclust/389093368727786622/clust --threads 22 --compressed 0 -v 3 \n",
      "\n",
      "Clustering step 1\n",
      "[=================================================================] 2.14K 0s 4ms\n",
      "Clustering step 2\n",
      "[=================================================================] 1.98K 0s 17ms\n",
      "Write merged clustering\n",
      "[=================================================================] 13.72K 0s 25ms\n",
      "Time for merging to clu_redundancy: 0h 0m 0s 1ms\n",
      "Time for processing: 0h 0m 0s 36ms\n",
      "createsubdb ./mmseq_db/tmpDir/17033438874298005812/clu_redundancy ./mmseq_db/inputDB ./mmseq_db/tmpDir/17033438874298005812/input_step_redundancy -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step_redundancy: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 1ms\n",
      "prefilter ./mmseq_db/tmpDir/17033438874298005812/input_step_redundancy ./mmseq_db/tmpDir/17033438874298005812/input_step_redundancy ./mmseq_db/tmpDir/17033438874298005812/pref_step0 --sub-mat nucl:nucleotide.out,aa:blosum62.out --seed-sub-mat nucl:nucleotide.out,aa:VTML80.out -s 1 -k 0 --k-score 2147483647 --alph-size nucl:5,aa:21 --max-seq-len 65535 --max-seqs 20 --split 0 --split-mode 2 --split-memory-limit 0 -c 0.8 --cov-mode 1 --comp-bias-corr 0 --diag-score 0 --exact-kmer-matching 0 --mask 1 --mask-lower-case 0 --min-ungapped-score 0 --add-self-matches 0 --spaced-kmer-mode 1 --db-load-mode 0 --pca 1 --pcb 1.5 --threads 22 --compressed 0 -v 3 \n",
      "\n",
      "Query database size: 1982 type: Aminoacid\n",
      "Estimated memory consumption: 983M\n",
      "Target database size: 1982 type: Aminoacid\n",
      "Index table k-mer threshold: 154 at k-mer size 6 \n",
      "Index table: counting k-mers\n",
      "[=================================================================] 1.98K 0s 28ms\n",
      "Index table: Masked residues: 2295\n",
      "Index table: fill\n",
      "[=================================================================] 1.98K 0s 15ms\n",
      "Index statistics\n",
      "Entries:          244855\n",
      "DB size:          489 MB\n",
      "Avg k-mer size:   0.003826\n",
      "Top 10 k-mers\n",
      "    GPLGGR\t7\n",
      "    GGDCDS\t7\n",
      "    WAGGAT\t6\n",
      "    FDGSWV\t6\n",
      "    EPWIGK\t5\n",
      "    SCGSPL\t5\n",
      "    WSGIEM\t5\n",
      "    SGVRHM\t5\n",
      "    DSQDGP\t5\n",
      "    VVCAGR\t5\n",
      "Time for index table init: 0h 0m 0s 415ms\n",
      "Process prefiltering step 1 of 1\n",
      "\n",
      "k-mer similarity threshold: 154\n",
      "Starting prefiltering scores calculation (step 1 of 1)\n",
      "Query db start 1 to 1982\n",
      "Target db start 1 to 1982\n",
      "[=================================================================] 1.98K 0s 12ms\n",
      "\n",
      "2.114947 k-mers per position\n",
      "130 DB matches per sequence\n",
      "0 overflows\n",
      "0 queries produce too many hits (truncated result)\n",
      "1 sequences passed prefiltering per query sequence\n",
      "1 median result list length\n",
      "0 sequences with 0 size result lists\n",
      "Time for merging to pref_step0: 0h 0m 0s 1ms\n",
      "Time for processing: 0h 0m 0s 723ms\n",
      "align ./mmseq_db/tmpDir/17033438874298005812/input_step_redundancy ./mmseq_db/tmpDir/17033438874298005812/input_step_redundancy ./mmseq_db/tmpDir/17033438874298005812/pref_step0 ./mmseq_db/tmpDir/17033438874298005812/aln_step0 --sub-mat nucl:nucleotide.out,aa:blosum62.out -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.3 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 1 --max-seq-len 65535 --comp-bias-corr 0 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca 1 --pcb 1.5 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --gap-open nucl:5,aa:11 --gap-extend nucl:2,aa:1 --zdrop 40 --threads 22 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 1982 type: Aminoacid\n",
      "Target database size: 1982 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 1.98K 0s 27ms\n",
      "Time for merging to aln_step0: 0h 0m 0s 1ms\n",
      "3011 alignments calculated\n",
      "2883 sequence pairs passed the thresholds (0.957489 of overall calculated)\n",
      "1.454591 hits per query sequence\n",
      "Time for processing: 0h 0m 0s 34ms\n",
      "clust ./mmseq_db/tmpDir/17033438874298005812/input_step_redundancy ./mmseq_db/tmpDir/17033438874298005812/aln_step0 ./mmseq_db/tmpDir/17033438874298005812/clu_step0 --cluster-mode 3 --max-iterations 1000 --similarity-type 2 --threads 22 --compressed 0 -v 3 \n",
      "\n",
      "Clustering mode: Greedy Low Mem\n",
      "Total time: 0h 0m 0s 2ms\n",
      "\n",
      "Size of the sequence database: 1982\n",
      "Size of the alignment database: 1982\n",
      "Number of clusters: 1680\n",
      "\n",
      "Writing results 0h 0m 0s 0ms\n",
      "Time for merging to clu_step0: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 4ms\n",
      "createsubdb ./mmseq_db/tmpDir/17033438874298005812/clu_step0 ./mmseq_db/tmpDir/17033438874298005812/input_step_redundancy ./mmseq_db/tmpDir/17033438874298005812/input_step1 -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step1: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "prefilter ./mmseq_db/tmpDir/17033438874298005812/input_step1 ./mmseq_db/tmpDir/17033438874298005812/input_step1 ./mmseq_db/tmpDir/17033438874298005812/pref_step1 --sub-mat nucl:nucleotide.out,aa:blosum62.out --seed-sub-mat nucl:nucleotide.out,aa:VTML80.out -s 3 -k 0 --k-score 2147483647 --alph-size nucl:5,aa:21 --max-seq-len 65535 --max-seqs 20 --split 0 --split-mode 2 --split-memory-limit 0 -c 0.8 --cov-mode 1 --comp-bias-corr 1 --diag-score 1 --exact-kmer-matching 0 --mask 1 --mask-lower-case 0 --min-ungapped-score 15 --add-self-matches 0 --spaced-kmer-mode 1 --db-load-mode 0 --pca 1 --pcb 1.5 --threads 22 --compressed 0 -v 3 \n",
      "\n",
      "Query database size: 1680 type: Aminoacid\n",
      "Estimated memory consumption: 982M\n",
      "Target database size: 1680 type: Aminoacid\n",
      "Index table k-mer threshold: 136 at k-mer size 6 \n",
      "Index table: counting k-mers\n",
      "[=================================================================] 1.68K 0s 22ms\n",
      "Index table: Masked residues: 2131\n",
      "Index table: fill\n",
      "[=================================================================] 1.68K 0s 14ms\n",
      "Index statistics\n",
      "Entries:          401901\n",
      "DB size:          490 MB\n",
      "Avg k-mer size:   0.006280\n",
      "Top 10 k-mers\n",
      "    SCGSPL\t5\n",
      "    GPLGGR\t5\n",
      "    TNYYYE\t4\n",
      "    VLNIKG\t4\n",
      "    SSLPSH\t4\n",
      "    LIGGGL\t4\n",
      "    LGYGPL\t4\n",
      "    DFEPYL\t4\n",
      "    SGVRHM\t4\n",
      "    RALTLP\t4\n",
      "Time for index table init: 0h 0m 0s 385ms\n",
      "Process prefiltering step 1 of 1\n",
      "\n",
      "k-mer similarity threshold: 136\n",
      "Starting prefiltering scores calculation (step 1 of 1)\n",
      "Query db start 1 to 1680\n",
      "Target db start 1 to 1680\n",
      "[=================================================================] 1.68K 0s 51ms\n",
      "\n",
      "23.932028 k-mers per position\n",
      "260 DB matches per sequence\n",
      "0 overflows\n",
      "0 queries produce too many hits (truncated result)\n",
      "1 sequences passed prefiltering per query sequence\n",
      "1 median result list length\n",
      "0 sequences with 0 size result lists\n",
      "Time for merging to pref_step1: 0h 0m 0s 1ms\n",
      "Time for processing: 0h 0m 0s 726ms\n",
      "align ./mmseq_db/tmpDir/17033438874298005812/input_step1 ./mmseq_db/tmpDir/17033438874298005812/input_step1 ./mmseq_db/tmpDir/17033438874298005812/pref_step1 ./mmseq_db/tmpDir/17033438874298005812/aln_step1 --sub-mat nucl:nucleotide.out,aa:blosum62.out -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.3 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 1 --max-seq-len 65535 --comp-bias-corr 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca 1 --pcb 1.5 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --gap-open nucl:5,aa:11 --gap-extend nucl:2,aa:1 --zdrop 40 --threads 22 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 1680 type: Aminoacid\n",
      "Target database size: 1680 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 1.68K 0s 23ms\n",
      "Time for merging to aln_step1: 0h 0m 0s 1ms\n",
      "2466 alignments calculated\n",
      "2058 sequence pairs passed the thresholds (0.834550 of overall calculated)\n",
      "1.225000 hits per query sequence\n",
      "Time for processing: 0h 0m 0s 32ms\n",
      "clust ./mmseq_db/tmpDir/17033438874298005812/input_step1 ./mmseq_db/tmpDir/17033438874298005812/aln_step1 ./mmseq_db/tmpDir/17033438874298005812/clu_step1 --cluster-mode 3 --max-iterations 1000 --similarity-type 2 --threads 22 --compressed 0 -v 3 \n",
      "\n",
      "Clustering mode: Greedy Low Mem\n",
      "Total time: 0h 0m 0s 4ms\n",
      "\n",
      "Size of the sequence database: 1680\n",
      "Size of the alignment database: 1680\n",
      "Number of clusters: 1544\n",
      "\n",
      "Writing results 0h 0m 0s 0ms\n",
      "Time for merging to clu_step1: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 5ms\n",
      "createsubdb ./mmseq_db/tmpDir/17033438874298005812/clu_step1 ./mmseq_db/tmpDir/17033438874298005812/input_step1 ./mmseq_db/tmpDir/17033438874298005812/input_step2 -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step2: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "prefilter ./mmseq_db/tmpDir/17033438874298005812/input_step2 ./mmseq_db/tmpDir/17033438874298005812/input_step2 ./mmseq_db/tmpDir/17033438874298005812/pref_step2 --sub-mat nucl:nucleotide.out,aa:blosum62.out --seed-sub-mat nucl:nucleotide.out,aa:VTML80.out -s 5 -k 0 --k-score 2147483647 --alph-size nucl:5,aa:21 --max-seq-len 65535 --max-seqs 20 --split 0 --split-mode 2 --split-memory-limit 0 -c 0.8 --cov-mode 1 --comp-bias-corr 1 --diag-score 1 --exact-kmer-matching 0 --mask 1 --mask-lower-case 0 --min-ungapped-score 15 --add-self-matches 0 --spaced-kmer-mode 1 --db-load-mode 0 --pca 1 --pcb 1.5 --threads 22 --compressed 0 -v 3 \n",
      "\n",
      "Query database size: 1544 type: Aminoacid\n",
      "Estimated memory consumption: 982M\n",
      "Target database size: 1544 type: Aminoacid\n",
      "Index table k-mer threshold: 118 at k-mer size 6 \n",
      "Index table: counting k-mers\n",
      "[=================================================================] 1.54K 0s 13ms\n",
      "Index table: Masked residues: 2096\n",
      "Index table: fill\n",
      "[=================================================================] 1.54K 0s 12ms\n",
      "Index statistics\n",
      "Entries:          389289\n",
      "DB size:          490 MB\n",
      "Avg k-mer size:   0.006083\n",
      "Top 10 k-mers\n",
      "    AATAAA\t4\n",
      "    EIIFKD\t4\n",
      "    LTGGLE\t4\n",
      "    TNYYYE\t4\n",
      "    SSLPSH\t4\n",
      "    LIGGGL\t4\n",
      "    LGYGPL\t4\n",
      "    DFEPYL\t4\n",
      "    SGVRHM\t4\n",
      "    RALTLP\t4\n",
      "Time for index table init: 0h 0m 0s 372ms\n",
      "Process prefiltering step 1 of 1\n",
      "\n",
      "k-mer similarity threshold: 118\n",
      "Starting prefiltering scores calculation (step 1 of 1)\n",
      "Query db start 1 to 1544\n",
      "Target db start 1 to 1544\n",
      "[=================================================================] 1.54K 0s 192ms\n",
      "\n",
      "179.745106 k-mers per position\n",
      "523 DB matches per sequence\n",
      "0 overflows\n",
      "0 queries produce too many hits (truncated result)\n",
      "3 sequences passed prefiltering per query sequence\n",
      "3 median result list length\n",
      "0 sequences with 0 size result lists\n",
      "Time for merging to pref_step2: 0h 0m 0s 1ms\n",
      "Time for processing: 0h 0m 0s 873ms\n",
      "align ./mmseq_db/tmpDir/17033438874298005812/input_step2 ./mmseq_db/tmpDir/17033438874298005812/input_step2 ./mmseq_db/tmpDir/17033438874298005812/pref_step2 ./mmseq_db/tmpDir/17033438874298005812/aln_step2 --sub-mat nucl:nucleotide.out,aa:blosum62.out -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.3 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 1 --max-seq-len 65535 --comp-bias-corr 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca 1 --pcb 1.5 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --gap-open nucl:5,aa:11 --gap-extend nucl:2,aa:1 --zdrop 40 --threads 22 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 1544 type: Aminoacid\n",
      "Target database size: 1544 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 1.54K 0s 19ms\n",
      "Time for merging to aln_step2: 0h 0m 0s 1ms\n",
      "3939 alignments calculated\n",
      "1729 sequence pairs passed the thresholds (0.438944 of overall calculated)\n",
      "1.119819 hits per query sequence\n",
      "Time for processing: 0h 0m 0s 24ms\n",
      "clust ./mmseq_db/tmpDir/17033438874298005812/input_step2 ./mmseq_db/tmpDir/17033438874298005812/aln_step2 ./mmseq_db/tmpDir/17033438874298005812/clu_step2 --cluster-mode 3 --max-iterations 1000 --similarity-type 2 --threads 22 --compressed 0 -v 3 \n",
      "\n",
      "Clustering mode: Greedy Low Mem\n",
      "Total time: 0h 0m 0s 18ms\n",
      "\n",
      "Size of the sequence database: 1544\n",
      "Size of the alignment database: 1544\n",
      "Number of clusters: 1460\n",
      "\n",
      "Writing results 0h 0m 0s 0ms\n",
      "Time for merging to clu_step2: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 20ms\n",
      "mergeclusters ./mmseq_db/inputDB ./mmseq_db/clusterDB ./mmseq_db/tmpDir/17033438874298005812/clu_redundancy ./mmseq_db/tmpDir/17033438874298005812/clu_step0 ./mmseq_db/tmpDir/17033438874298005812/clu_step1 ./mmseq_db/tmpDir/17033438874298005812/clu_step2 --threads 22 --compressed 0 -v 3 \n",
      "\n",
      "Clustering step 1\n",
      "[=================================================================] 1.98K 0s 15ms\n",
      "Clustering step 2\n",
      "[=================================================================] 1.68K 0s 25ms\n",
      "Clustering step 3\n",
      "[=================================================================] 1.54K 0s 35ms\n",
      "Clustering step 4\n",
      "[=================================================================] 1.46K 0s 49ms\n",
      "Write merged clustering\n",
      "[=================================================================] 13.72K 0s 56ms\n",
      "Time for merging to clusterDB: 0h 0m 0s 1ms\n",
      "Time for processing: 0h 0m 0s 62ms\n",
      "result2repseq ./mmseq_db/inputDB ./mmseq_db/clusterDB ./mmseq_db/repDB \n",
      "\n",
      "MMseqs Version:\t13.45111\n",
      "Preload mode\t0\n",
      "Compressed  \t0\n",
      "Threads     \t22\n",
      "Verbosity   \t3\n",
      "\n",
      "[=================================================================] 1.46K 0s 4ms\n",
      "Time for merging to repDB: 0h 0m 0s 1ms\n",
      "Time for processing: 0h 0m 0s 14ms\n",
      "convert2fasta ./mmseq_db/repDB ./fasta/high_resolution_2dot8_protein_sequences_before_2000_representatives.fasta \n",
      "\n",
      "MMseqs Version:\t13.45111\n",
      "Use header DB\tfalse\n",
      "Verbosity    \t3\n",
      "\n",
      "Start writing file to ./fasta/high_resolution_2dot8_protein_sequences_before_2000_representatives.fasta\n",
      "Time for processing: 0h 0m 0s 1ms\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fasta_path = './fasta/high_resolution_2dot8_protein_sequences_before_2000'\n",
    "# Ensure tmpDir exists\n",
    "os.makedirs('./mmseq_db/tmpDir', exist_ok=True)\n",
    "\n",
    "# Step 1: create sequence DB\n",
    "os.system(f'mmseqs createdb {fasta_path}.fasta ./mmseq_db/inputDB')\n",
    "\n",
    "# Step 2: cluster sequences\n",
    "os.system('mmseqs cluster ./mmseq_db/inputDB ./mmseq_db/clusterDB ./mmseq_db/tmpDir --min-seq-id 0.3 -c 0.8 --cov-mode 1')\n",
    "\n",
    "# Step 3: extract representative sequences\n",
    "os.system('mmseqs result2repseq ./mmseq_db/inputDB ./mmseq_db/clusterDB ./mmseq_db/repDB')\n",
    "\n",
    "# Step 4: convert the repDB to FASTA\n",
    "os.system(f'mmseqs convert2fasta ./mmseq_db/repDB {fasta_path}_representatives.fasta')\n",
    "\n",
    "# Clear mmseq_db dir\n",
    "os.system('rm mmseq_db/* -rf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "7f87a559-b4e2-4a99-8d4e-3a3bdbc3160e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1460 1460\n"
     ]
    }
   ],
   "source": [
    "with open(f'{fasta_path}_representatives.fasta', 'r') as f:\n",
    "    pdb_chain = []\n",
    "    for line in f.readlines():\n",
    "        if line[0]=='>':\n",
    "            pdb_chain.append(line[1:])\n",
    "pdb_chain_arr = np.array([i.split('_') for i in pdb_chain])\n",
    "print(len(pdb_chain_arr), len(np.unique(pdb_chain_arr[:,0])))\n",
    "#np.save('mem_prot_rep_2dot8.npy', pdb_chain_arr)\n",
    "np.save('prot_rep_2dot8_before_2000.npy', pdb_chain_arr)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.23"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
