{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "120ef59f-945a-4fe5-9337-994e062657ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import csv\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "c9f8578c-6ee9-480e-bd17-28f57b3cbc99",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('./data/high_resolution_2_membrane_protein_sequences_after_2000.csv', newline='') as f:\n",
    "    reader = csv.reader(f)\n",
    "    file = list(reader)  # convert the reader to a list of rows\n",
    "membrane_pdb_ids = []\n",
    "for line in file[2:]:\n",
    "    if len(line[0])==4:\n",
    "        pdb_id = line[0]\n",
    "        membrane_pdb_ids.append(pdb_id)\n",
    "    elif len(line[0])>1:\n",
    "        pdb_id = line[0][:2]+line[0][-2:]\n",
    "        membrane_pdb_ids.append(pdb_id)\n",
    "        \n",
    "with open('./data/high_resolution_2_protein_sequences_after_2000.csv', newline='') as f:\n",
    "    reader = csv.reader(f)\n",
    "    file = list(reader)  # convert the reader to a list of rows"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "5688d3b6-ecc2-40de-8077-8701a1ef6bd5",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "fasta = []\n",
    "names = []\n",
    "for line in file[2:]:\n",
    "    if len(line[0])==4:\n",
    "        pdb_id = line[0]\n",
    "    elif len(line[0])>1:\n",
    "        pdb_id = line[0][:2]+line[0][-2:]\n",
    "    if not len(line[2]) or line[0] == 'Entry ID' or line[-2] != 'polypeptide(L)' or pdb_id in membrane_pdb_ids:\n",
    "        continue\n",
    "    #line = line.split(',')\n",
    "    chains = [i.strip() for i in line[2].split(',')]\n",
    "    if len(chains)-1:\n",
    "        for c in chains:\n",
    "            fasta.append(f'>{pdb_id}_{c}')\n",
    "            names.append(f'>{pdb_id}_{c}')\n",
    "            fasta.append(line[6])\n",
    "    else:\n",
    "        fasta.append(f'>{pdb_id}_{chains[0]}')\n",
    "        names.append(f'>{pdb_id}_{chains[0]}')\n",
    "        fasta.append(line[6])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "20fbfb0e-4238-437f-977c-9d453585c9eb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "196062 196062\n"
     ]
    }
   ],
   "source": [
    "un, ind, counts = np.unique(names, return_counts=True, return_index=True)\n",
    "print(len(names), len(un))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "63429e97-30da-4ddb-bc5f-2b09366719b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('./fasta/high_resolution_2_protein_sequences_after_2000.fasta', 'w') as f:\n",
    "    for line in fasta:\n",
    "        f.write(f'{line}\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "6983d69f-279e-4669-9e49-f81978f97da9",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "createdb ./fasta/high_resolution_2_protein_sequences_after_2000.fasta ./mmseq_db/inputDB \n",
      "\n",
      "MMseqs Version:       \t13.45111\n",
      "Database type         \t0\n",
      "Shuffle input database\ttrue\n",
      "Createdb mode         \t0\n",
      "Write lookup file     \t1\n",
      "Offset of numeric ids \t0\n",
      "Compressed            \t0\n",
      "Verbosity             \t3\n",
      "\n",
      "Converting sequences\n",
      "[===================\n",
      "Time for merging to inputDB_h: 0h 0m 0s 15ms\n",
      "Time for merging to inputDB: 0h 0m 0s 30ms\n",
      "Database type: Aminoacid\n",
      "Time for processing: 0h 0m 0s 398ms\n",
      "cluster ./mmseq_db/inputDB ./mmseq_db/clusterDB ./mmseq_db/tmpDir --min-seq-id 0.3 -c 0.8 --cov-mode 1 \n",
      "\n",
      "MMseqs Version:                     \t13.45111\n",
      "Substitution matrix                 \tnucl:nucleotide.out,aa:blosum62.out\n",
      "Seed substitution matrix            \tnucl:nucleotide.out,aa:VTML80.out\n",
      "Sensitivity                         \t4\n",
      "k-mer length                        \t0\n",
      "k-score                             \t2147483647\n",
      "Alphabet size                       \tnucl:5,aa:21\n",
      "Max sequence length                 \t65535\n",
      "Max results per query               \t20\n",
      "Split database                      \t0\n",
      "Split mode                          \t2\n",
      "Split memory limit                  \t0\n",
      "Coverage threshold                  \t0.8\n",
      "Coverage mode                       \t1\n",
      "Compositional bias                  \t1\n",
      "Diagonal scoring                    \ttrue\n",
      "Exact k-mer matching                \t0\n",
      "Mask residues                       \t1\n",
      "Mask lower case residues            \t0\n",
      "Minimum diagonal score              \t15\n",
      "Include identical seq. id.          \tfalse\n",
      "Spaced k-mers                       \t1\n",
      "Preload mode                        \t0\n",
      "Pseudo count a                      \t1\n",
      "Pseudo count b                      \t1.5\n",
      "Spaced k-mer pattern                \t\n",
      "Local temporary path                \t\n",
      "Threads                             \t22\n",
      "Compressed                          \t0\n",
      "Verbosity                           \t3\n",
      "Add backtrace                       \tfalse\n",
      "Alignment mode                      \t3\n",
      "Alignment mode                      \t0\n",
      "Allow wrapped scoring               \tfalse\n",
      "E-value threshold                   \t0.001\n",
      "Seq. id. threshold                  \t0.3\n",
      "Min alignment length                \t0\n",
      "Seq. id. mode                       \t0\n",
      "Alternative alignments              \t0\n",
      "Max reject                          \t2147483647\n",
      "Max accept                          \t2147483647\n",
      "Score bias                          \t0\n",
      "Realign hits                        \tfalse\n",
      "Realign score bias                  \t-0.2\n",
      "Realign max seqs                    \t2147483647\n",
      "Gap open cost                       \tnucl:5,aa:11\n",
      "Gap extension cost                  \tnucl:2,aa:1\n",
      "Zdrop                               \t40\n",
      "Rescore mode                        \t0\n",
      "Remove hits by seq. id. and coverage\tfalse\n",
      "Sort results                        \t0\n",
      "Cluster mode                        \t0\n",
      "Max connected component depth       \t1000\n",
      "Similarity type                     \t2\n",
      "Single step clustering              \tfalse\n",
      "Cascaded clustering steps           \t3\n",
      "Cluster reassign                    \tfalse\n",
      "Remove temporary files              \tfalse\n",
      "Force restart with latest tmp       \tfalse\n",
      "MPI runner                          \t\n",
      "k-mers per sequence                 \t21\n",
      "Scale k-mers per sequence           \tnucl:0.200,aa:0.000\n",
      "Adjust k-mer length                 \tfalse\n",
      "Shift hash                          \t67\n",
      "Include only extendable             \tfalse\n",
      "Skip repeating k-mers               \tfalse\n",
      "\n",
      "Set cluster sensitivity to -s 5.000000\n",
      "Set cluster mode GREEDY MEM\n",
      "Set cluster iterations to 3\n",
      "linclust ./mmseq_db/inputDB ./mmseq_db/tmpDir/14959533115294637411/clu_redundancy ./mmseq_db/tmpDir/14959533115294637411/linclust --cluster-mode 3 --max-iterations 1000 --similarity-type 2 --threads 22 --compressed 0 -v 3 --sub-mat nucl:nucleotide.out,aa:blosum62.out -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.3 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 1 --max-seq-len 65535 --comp-bias-corr 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca 1 --pcb 1.5 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --gap-open nucl:5,aa:11 --gap-extend nucl:2,aa:1 --zdrop 40 --alph-size nucl:5,aa:13 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale nucl:0.200,aa:0.000 --adjust-kmer-len 0 --mask 0 --mask-lower-case 0 -k 0 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --rescore-mode 0 --filter-hits 0 --sort-results 0 --remove-tmp-files 0 --force-reuse 0 \n",
      "\n",
      "kmermatcher ./mmseq_db/inputDB ./mmseq_db/tmpDir/14959533115294637411/linclust/7589472165642968301/pref --sub-mat nucl:nucleotide.out,aa:blosum62.out --alph-size nucl:5,aa:13 --min-seq-id 0.3 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale nucl:0.200,aa:0.000 --adjust-kmer-len 0 --mask 0 --mask-lower-case 0 --cov-mode 1 -k 0 -c 0.8 --max-seq-len 65535 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --threads 22 --compressed 0 -v 3 \n",
      "\n",
      "kmermatcher ./mmseq_db/inputDB ./mmseq_db/tmpDir/14959533115294637411/linclust/7589472165642968301/pref --sub-mat nucl:nucleotide.out,aa:blosum62.out --alph-size nucl:5,aa:13 --min-seq-id 0.3 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale nucl:0.200,aa:0.000 --adjust-kmer-len 0 --mask 0 --mask-lower-case 0 --cov-mode 1 -k 0 -c 0.8 --max-seq-len 65535 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --threads 22 --compressed 0 -v 3 \n",
      "\n",
      "Database size: 196062 type: Aminoacid\n",
      "Reduced amino acid alphabet: (A S T) (C) (D B N) (E Q Z) (F Y) (G) (H) (I V) (K R) (L J M) (P) (W) (X) \n",
      "\n",
      "Generate k-mers list for 1 split\n",
      "[=================================================================] 196.06K 0s 163ms\n",
      "Sort kmer 0h 0m 0s 105ms\n",
      "Sort by rep. sequence 0h 0m 0s 87ms\n",
      "Time for fill: 0h 0m 0s 31ms\n",
      "Time for merging to pref: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 478ms\n",
      "rescorediagonal ./mmseq_db/inputDB ./mmseq_db/inputDB ./mmseq_db/tmpDir/14959533115294637411/linclust/7589472165642968301/pref ./mmseq_db/tmpDir/14959533115294637411/linclust/7589472165642968301/pref_rescore1 --sub-mat nucl:nucleotide.out,aa:blosum62.out --rescore-mode 0 --wrapped-scoring 0 --filter-hits 0 -e 0.001 -c 0.8 -a 0 --cov-mode 1 --min-seq-id 0.5 --min-aln-len 0 --seq-id-mode 0 --add-self-matches 0 --sort-results 0 --db-load-mode 0 --threads 22 --compressed 0 -v 3 \n",
      "\n",
      "[=================================================================] 196.06K 0s 43ms\n",
      "Time for merging to pref_rescore1: 0h 0m 0s 82ms\n",
      "Time for processing: 0h 0m 0s 252ms\n",
      "clust ./mmseq_db/inputDB ./mmseq_db/tmpDir/14959533115294637411/linclust/7589472165642968301/pref_rescore1 ./mmseq_db/tmpDir/14959533115294637411/linclust/7589472165642968301/pre_clust --cluster-mode 3 --max-iterations 1000 --similarity-type 2 --threads 22 --compressed 0 -v 3 \n",
      "\n",
      "Clustering mode: Greedy Low Mem\n",
      "Total time: 0h 0m 0s 66ms\n",
      "\n",
      "Size of the sequence database: 196062\n",
      "Size of the alignment database: 196062\n",
      "Number of clusters: 28251\n",
      "\n",
      "Writing results 0h 0m 0s 6ms\n",
      "Time for merging to pre_clust: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 144ms\n",
      "createsubdb ./mmseq_db/tmpDir/14959533115294637411/linclust/7589472165642968301/order_redundancy ./mmseq_db/inputDB ./mmseq_db/tmpDir/14959533115294637411/linclust/7589472165642968301/input_step_redundancy -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step_redundancy: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 11ms\n",
      "createsubdb ./mmseq_db/tmpDir/14959533115294637411/linclust/7589472165642968301/order_redundancy ./mmseq_db/tmpDir/14959533115294637411/linclust/7589472165642968301/pref ./mmseq_db/tmpDir/14959533115294637411/linclust/7589472165642968301/pref_filter1 -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to pref_filter1: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 72ms\n",
      "filterdb ./mmseq_db/tmpDir/14959533115294637411/linclust/7589472165642968301/pref_filter1 ./mmseq_db/tmpDir/14959533115294637411/linclust/7589472165642968301/pref_filter2 --filter-file ./mmseq_db/tmpDir/14959533115294637411/linclust/7589472165642968301/order_redundancy --threads 22 --compressed 0 -v 3 \n",
      "\n",
      "Filtering using file(s)\n",
      "[=================================================================] 28.25K 0s 13ms\n",
      "Time for merging to pref_filter2: 0h 0m 0s 7ms\n",
      "Time for processing: 0h 0m 0s 132ms\n",
      "rescorediagonal ./mmseq_db/tmpDir/14959533115294637411/linclust/7589472165642968301/input_step_redundancy ./mmseq_db/tmpDir/14959533115294637411/linclust/7589472165642968301/input_step_redundancy ./mmseq_db/tmpDir/14959533115294637411/linclust/7589472165642968301/pref_filter2 ./mmseq_db/tmpDir/14959533115294637411/linclust/7589472165642968301/pref_rescore2 --sub-mat nucl:nucleotide.out,aa:blosum62.out --rescore-mode 1 --wrapped-scoring 0 --filter-hits 1 -e 0.001 -c 0.8 -a 0 --cov-mode 1 --min-seq-id 0.3 --min-aln-len 0 --seq-id-mode 0 --add-self-matches 0 --sort-results 0 --db-load-mode 0 --threads 22 --compressed 0 -v 3 \n",
      "\n",
      "[=================================================================] 28.25K 0s 14ms\n",
      "Time for merging to pref_rescore2: 0h 0m 0s 10ms\n",
      "Time for processing: 0h 0m 0s 79ms\n",
      "align ./mmseq_db/tmpDir/14959533115294637411/linclust/7589472165642968301/input_step_redundancy ./mmseq_db/tmpDir/14959533115294637411/linclust/7589472165642968301/input_step_redundancy ./mmseq_db/tmpDir/14959533115294637411/linclust/7589472165642968301/pref_rescore2 ./mmseq_db/tmpDir/14959533115294637411/linclust/7589472165642968301/aln --sub-mat nucl:nucleotide.out,aa:blosum62.out -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.3 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 1 --max-seq-len 65535 --comp-bias-corr 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca 1 --pcb 1.5 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --gap-open nucl:5,aa:11 --gap-extend nucl:2,aa:1 --zdrop 40 --threads 22 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 28251 type: Aminoacid\n",
      "Target database size: 28251 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 28.25K 0s 190ms\n",
      "Time for merging to aln: 0h 0m 0s 10ms\n",
      "35554 alignments calculated\n",
      "32941 sequence pairs passed the thresholds (0.926506 of overall calculated)\n",
      "1.166012 hits per query sequence\n",
      "Time for processing: 0h 0m 0s 247ms\n",
      "clust ./mmseq_db/tmpDir/14959533115294637411/linclust/7589472165642968301/input_step_redundancy ./mmseq_db/tmpDir/14959533115294637411/linclust/7589472165642968301/aln ./mmseq_db/tmpDir/14959533115294637411/linclust/7589472165642968301/clust --cluster-mode 3 --max-iterations 1000 --similarity-type 2 --threads 22 --compressed 0 -v 3 \n",
      "\n",
      "Clustering mode: Greedy Low Mem\n",
      "Total time: 0h 0m 0s 46ms\n",
      "\n",
      "Size of the sequence database: 28251\n",
      "Size of the alignment database: 28251\n",
      "Number of clusters: 25539\n",
      "\n",
      "Writing results 0h 0m 0s 3ms\n",
      "Time for merging to clust: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 95ms\n",
      "mergeclusters ./mmseq_db/inputDB ./mmseq_db/tmpDir/14959533115294637411/clu_redundancy ./mmseq_db/tmpDir/14959533115294637411/linclust/7589472165642968301/pre_clust ./mmseq_db/tmpDir/14959533115294637411/linclust/7589472165642968301/clust --threads 22 --compressed 0 -v 3 \n",
      "\n",
      "Clustering step 1\n",
      "[=================================================================] 28.25K 0s 10ms\n",
      "Clustering step 2\n",
      "[=================================================================] 25.54K 0s 28ms\n",
      "Write merged clustering\n",
      "[=================================================================] 196.06K 0s 51ms\n",
      "Time for merging to clu_redundancy: 0h 0m 0s 9ms\n",
      "Time for processing: 0h 0m 0s 84ms\n",
      "createsubdb ./mmseq_db/tmpDir/14959533115294637411/clu_redundancy ./mmseq_db/inputDB ./mmseq_db/tmpDir/14959533115294637411/input_step_redundancy -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step_redundancy: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 11ms\n",
      "prefilter ./mmseq_db/tmpDir/14959533115294637411/input_step_redundancy ./mmseq_db/tmpDir/14959533115294637411/input_step_redundancy ./mmseq_db/tmpDir/14959533115294637411/pref_step0 --sub-mat nucl:nucleotide.out,aa:blosum62.out --seed-sub-mat nucl:nucleotide.out,aa:VTML80.out -s 1 -k 0 --k-score 2147483647 --alph-size nucl:5,aa:21 --max-seq-len 65535 --max-seqs 20 --split 0 --split-mode 2 --split-memory-limit 0 -c 0.8 --cov-mode 1 --comp-bias-corr 0 --diag-score 0 --exact-kmer-matching 0 --mask 1 --mask-lower-case 0 --min-ungapped-score 0 --add-self-matches 0 --spaced-kmer-mode 1 --db-load-mode 0 --pca 1 --pcb 1.5 --threads 22 --compressed 0 -v 3 \n",
      "\n",
      "Query database size: 25539 type: Aminoacid\n",
      "Estimated memory consumption: 1G\n",
      "Target database size: 25539 type: Aminoacid\n",
      "Index table k-mer threshold: 154 at k-mer size 6 \n",
      "Index table: counting k-mers\n",
      "[=================================================================] 25.54K 0s 95ms\n",
      "Index table: Masked residues: 81528\n",
      "Index table: fill\n",
      "[=================================================================] 25.54K 0s 45ms\n",
      "Index statistics\n",
      "Entries:          2679313\n",
      "DB size:          503 MB\n",
      "Avg k-mer size:   0.041864\n",
      "Top 10 k-mers\n",
      "    SGVRHM\t1147\n",
      "    DLTNFQ\t392\n",
      "    SMGQGR\t202\n",
      "    SSRNFQ\t155\n",
      "    TGQMGS\t142\n",
      "    PRSMMT\t125\n",
      "    YDPTLY\t105\n",
      "    GLPGMN\t46\n",
      "    GLPGMP\t45\n",
      "    GLPGMQ\t40\n",
      "Time for index table init: 0h 0m 0s 379ms\n",
      "Process prefiltering step 1 of 1\n",
      "\n",
      "k-mer similarity threshold: 154\n",
      "Starting prefiltering scores calculation (step 1 of 1)\n",
      "Query db start 1 to 25539\n",
      "Target db start 1 to 25539\n",
      "[=================================================================] 25.54K 0s 117ms\n",
      "\n",
      "1.905179 k-mers per position\n",
      "227 DB matches per sequence\n",
      "0 overflows\n",
      "0 queries produce too many hits (truncated result)\n",
      "3 sequences passed prefiltering per query sequence\n",
      "1 median result list length\n",
      "0 sequences with 0 size result lists\n",
      "Time for merging to pref_step0: 0h 0m 0s 8ms\n",
      "Time for processing: 0h 0m 0s 813ms\n",
      "align ./mmseq_db/tmpDir/14959533115294637411/input_step_redundancy ./mmseq_db/tmpDir/14959533115294637411/input_step_redundancy ./mmseq_db/tmpDir/14959533115294637411/pref_step0 ./mmseq_db/tmpDir/14959533115294637411/aln_step0 --sub-mat nucl:nucleotide.out,aa:blosum62.out -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.3 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 1 --max-seq-len 65535 --comp-bias-corr 0 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca 1 --pcb 1.5 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --gap-open nucl:5,aa:11 --gap-extend nucl:2,aa:1 --zdrop 40 --threads 22 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 25539 type: Aminoacid\n",
      "Target database size: 25539 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 25.54K 0s 528ms\n",
      "Time for merging to aln_step0: 0h 0m 0s 7ms\n",
      "74528 alignments calculated\n",
      "39750 sequence pairs passed the thresholds (0.533357 of overall calculated)\n",
      "1.556443 hits per query sequence\n",
      "Time for processing: 0h 0m 0s 608ms\n",
      "clust ./mmseq_db/tmpDir/14959533115294637411/input_step_redundancy ./mmseq_db/tmpDir/14959533115294637411/aln_step0 ./mmseq_db/tmpDir/14959533115294637411/clu_step0 --cluster-mode 3 --max-iterations 1000 --similarity-type 2 --threads 22 --compressed 0 -v 3 \n",
      "\n",
      "Clustering mode: Greedy Low Mem\n",
      "Total time: 0h 0m 0s 76ms\n",
      "\n",
      "Size of the sequence database: 25539\n",
      "Size of the alignment database: 25539\n",
      "Number of clusters: 21719\n",
      "\n",
      "Writing results 0h 0m 0s 3ms\n",
      "Time for merging to clu_step0: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 131ms\n",
      "createsubdb ./mmseq_db/tmpDir/14959533115294637411/clu_step0 ./mmseq_db/tmpDir/14959533115294637411/input_step_redundancy ./mmseq_db/tmpDir/14959533115294637411/input_step1 -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step1: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 4ms\n",
      "prefilter ./mmseq_db/tmpDir/14959533115294637411/input_step1 ./mmseq_db/tmpDir/14959533115294637411/input_step1 ./mmseq_db/tmpDir/14959533115294637411/pref_step1 --sub-mat nucl:nucleotide.out,aa:blosum62.out --seed-sub-mat nucl:nucleotide.out,aa:VTML80.out -s 3 -k 0 --k-score 2147483647 --alph-size nucl:5,aa:21 --max-seq-len 65535 --max-seqs 20 --split 0 --split-mode 2 --split-memory-limit 0 -c 0.8 --cov-mode 1 --comp-bias-corr 1 --diag-score 1 --exact-kmer-matching 0 --mask 1 --mask-lower-case 0 --min-ungapped-score 15 --add-self-matches 0 --spaced-kmer-mode 1 --db-load-mode 0 --pca 1 --pcb 1.5 --threads 22 --compressed 0 -v 3 \n",
      "\n",
      "Query database size: 21719 type: Aminoacid\n",
      "Estimated memory consumption: 1G\n",
      "Target database size: 21719 type: Aminoacid\n",
      "Index table k-mer threshold: 136 at k-mer size 6 \n",
      "Index table: counting k-mers\n",
      "[=================================================================] 21.72K 0s 83ms\n",
      "Index table: Masked residues: 71070\n",
      "Index table: fill\n",
      "[=================================================================] 21.72K 0s 60ms\n",
      "Index statistics\n",
      "Entries:          4224667\n",
      "DB size:          512 MB\n",
      "Avg k-mer size:   0.066010\n",
      "Top 10 k-mers\n",
      "    SSLPSH\t996\n",
      "    SGVRHM\t967\n",
      "    GVLTLY\t304\n",
      "    DLTNFQ\t302\n",
      "    SMGQGR\t167\n",
      "    SSRNFQ\t137\n",
      "    HMSTQQ\t122\n",
      "    TGQMGS\t118\n",
      "    PRSMMT\t111\n",
      "    YDPTLY\t82\n",
      "Time for index table init: 0h 0m 0s 381ms\n",
      "Process prefiltering step 1 of 1\n",
      "\n",
      "k-mer similarity threshold: 136\n",
      "Starting prefiltering scores calculation (step 1 of 1)\n",
      "Query db start 1 to 21719\n",
      "Target db start 1 to 21719\n",
      "[=================================================================] 21.72K 0s 475ms\n",
      "\n",
      "19.735218 k-mers per position\n",
      "591 DB matches per sequence\n",
      "0 overflows\n",
      "0 queries produce too many hits (truncated result)\n",
      "4 sequences passed prefiltering per query sequence\n",
      "3 median result list length\n",
      "0 sequences with 0 size result lists\n",
      "Time for merging to pref_step1: 0h 0m 0s 7ms\n",
      "Time for processing: 0h 0m 1s 132ms\n",
      "align ./mmseq_db/tmpDir/14959533115294637411/input_step1 ./mmseq_db/tmpDir/14959533115294637411/input_step1 ./mmseq_db/tmpDir/14959533115294637411/pref_step1 ./mmseq_db/tmpDir/14959533115294637411/aln_step1 --sub-mat nucl:nucleotide.out,aa:blosum62.out -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.3 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 1 --max-seq-len 65535 --comp-bias-corr 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca 1 --pcb 1.5 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --gap-open nucl:5,aa:11 --gap-extend nucl:2,aa:1 --zdrop 40 --threads 22 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 21719 type: Aminoacid\n",
      "Target database size: 21719 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 21.72K 0s 455ms\n",
      "Time for merging to aln_step1: 0h 0m 0s 5ms\n",
      "81238 alignments calculated\n",
      "30045 sequence pairs passed the thresholds (0.369839 of overall calculated)\n",
      "1.383351 hits per query sequence\n",
      "Time for processing: 0h 0m 0s 535ms\n",
      "clust ./mmseq_db/tmpDir/14959533115294637411/input_step1 ./mmseq_db/tmpDir/14959533115294637411/aln_step1 ./mmseq_db/tmpDir/14959533115294637411/clu_step1 --cluster-mode 3 --max-iterations 1000 --similarity-type 2 --threads 22 --compressed 0 -v 3 \n",
      "\n",
      "Clustering mode: Greedy Low Mem\n",
      "Total time: 0h 0m 0s 24ms\n",
      "\n",
      "Size of the sequence database: 21719\n",
      "Size of the alignment database: 21719\n",
      "Number of clusters: 19032\n",
      "\n",
      "Writing results 0h 0m 0s 6ms\n",
      "Time for merging to clu_step1: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 37ms\n",
      "createsubdb ./mmseq_db/tmpDir/14959533115294637411/clu_step1 ./mmseq_db/tmpDir/14959533115294637411/input_step1 ./mmseq_db/tmpDir/14959533115294637411/input_step2 -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step2: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 3ms\n",
      "prefilter ./mmseq_db/tmpDir/14959533115294637411/input_step2 ./mmseq_db/tmpDir/14959533115294637411/input_step2 ./mmseq_db/tmpDir/14959533115294637411/pref_step2 --sub-mat nucl:nucleotide.out,aa:blosum62.out --seed-sub-mat nucl:nucleotide.out,aa:VTML80.out -s 5 -k 0 --k-score 2147483647 --alph-size nucl:5,aa:21 --max-seq-len 65535 --max-seqs 20 --split 0 --split-mode 2 --split-memory-limit 0 -c 0.8 --cov-mode 1 --comp-bias-corr 1 --diag-score 1 --exact-kmer-matching 0 --mask 1 --mask-lower-case 0 --min-ungapped-score 15 --add-self-matches 0 --spaced-kmer-mode 1 --db-load-mode 0 --pca 1 --pcb 1.5 --threads 22 --compressed 0 -v 3 \n",
      "\n",
      "Query database size: 19032 type: Aminoacid\n",
      "Estimated memory consumption: 1G\n",
      "Target database size: 19032 type: Aminoacid\n",
      "Index table k-mer threshold: 118 at k-mer size 6 \n",
      "Index table: counting k-mers\n",
      "[=================================================================] 19.03K 0s 85ms\n",
      "Index table: Masked residues: 63064\n",
      "Index table: fill\n",
      "[=================================================================] 19.03K 0s 71ms\n",
      "Index statistics\n",
      "Entries:          3773235\n",
      "DB size:          509 MB\n",
      "Avg k-mer size:   0.058957\n",
      "Top 10 k-mers\n",
      "    SSLPSH\t856\n",
      "    SGVRHM\t834\n",
      "    SSVLEN\t254\n",
      "    GVLTLY\t254\n",
      "    LVRSAS\t198\n",
      "    SSRNFQ\t124\n",
      "    HMSTQQ\t115\n",
      "    TGQMGS\t108\n",
      "    PRSMMT\t104\n",
      "    YDPTLY\t78\n",
      "Time for index table init: 0h 0m 0s 464ms\n",
      "Process prefiltering step 1 of 1\n",
      "\n",
      "k-mer similarity threshold: 118\n",
      "Starting prefiltering scores calculation (step 1 of 1)\n",
      "Query db start 1 to 19032\n",
      "Target db start 1 to 19032\n",
      "[=================================================================] 19.03K 2s 272ms\n",
      "\n",
      "145.294446 k-mers per position\n",
      "2369 DB matches per sequence\n",
      "0 overflows\n",
      "0 queries produce too many hits (truncated result)\n",
      "12 sequences passed prefiltering per query sequence\n",
      "14 median result list length\n",
      "0 sequences with 0 size result lists\n",
      "Time for merging to pref_step2: 0h 0m 0s 13ms\n",
      "Time for processing: 0h 0m 3s 172ms\n",
      "align ./mmseq_db/tmpDir/14959533115294637411/input_step2 ./mmseq_db/tmpDir/14959533115294637411/input_step2 ./mmseq_db/tmpDir/14959533115294637411/pref_step2 ./mmseq_db/tmpDir/14959533115294637411/aln_step2 --sub-mat nucl:nucleotide.out,aa:blosum62.out -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.3 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 1 --max-seq-len 65535 --comp-bias-corr 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca 1 --pcb 1.5 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --gap-open nucl:5,aa:11 --gap-extend nucl:2,aa:1 --zdrop 40 --threads 22 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 19032 type: Aminoacid\n",
      "Target database size: 19032 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 19.03K 0s 884ms\n",
      "Time for merging to aln_step2: 0h 0m 0s 15ms\n",
      "142884 alignments calculated\n",
      "22375 sequence pairs passed the thresholds (0.156596 of overall calculated)\n",
      "1.175652 hits per query sequence\n",
      "Time for processing: 0h 0m 0s 964ms\n",
      "clust ./mmseq_db/tmpDir/14959533115294637411/input_step2 ./mmseq_db/tmpDir/14959533115294637411/aln_step2 ./mmseq_db/tmpDir/14959533115294637411/clu_step2 --cluster-mode 3 --max-iterations 1000 --similarity-type 2 --threads 22 --compressed 0 -v 3 \n",
      "\n",
      "Clustering mode: Greedy Low Mem\n",
      "Total time: 0h 0m 0s 24ms\n",
      "\n",
      "Size of the sequence database: 19032\n",
      "Size of the alignment database: 19032\n",
      "Number of clusters: 17704\n",
      "\n",
      "Writing results 0h 0m 0s 4ms\n",
      "Time for merging to clu_step2: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 37ms\n",
      "mergeclusters ./mmseq_db/inputDB ./mmseq_db/clusterDB ./mmseq_db/tmpDir/14959533115294637411/clu_redundancy ./mmseq_db/tmpDir/14959533115294637411/clu_step0 ./mmseq_db/tmpDir/14959533115294637411/clu_step1 ./mmseq_db/tmpDir/14959533115294637411/clu_step2 --threads 22 --compressed 0 -v 3 \n",
      "\n",
      "Clustering step 1\n",
      "[=================================================================] 25.54K 0s 10ms\n",
      "Clustering step 2\n",
      "[=================================================================] 21.72K 0s 28ms\n",
      "Clustering step 3\n",
      "[=================================================================] 19.03K 0s 45ms\n",
      "Clustering step 4\n",
      "[=================================================================] 17.70K 0s 63ms\n",
      "Write merged clustering\n",
      "[=================================================================] 196.06K 0s 95ms\n",
      "Time for merging to clusterDB: 0h 0m 0s 10ms\n",
      "Time for processing: 0h 0m 0s 176ms\n",
      "result2repseq ./mmseq_db/inputDB ./mmseq_db/clusterDB ./mmseq_db/repDB \n",
      "\n",
      "MMseqs Version:\t13.45111\n",
      "Preload mode\t0\n",
      "Compressed  \t0\n",
      "Threads     \t22\n",
      "Verbosity   \t3\n",
      "\n",
      "[=================================================================] 17.70K 0s 8ms\n",
      "Time for merging to repDB: 0h 0m 0s 12ms\n",
      "Time for processing: 0h 0m 0s 84ms\n",
      "./fasta/high_resolution_2_protein_sequences_after_2000_representatives.fasta exists and will be overwritten\n",
      "convert2fasta ./mmseq_db/repDB ./fasta/high_resolution_2_protein_sequences_after_2000_representatives.fasta \n",
      "\n",
      "MMseqs Version:\t13.45111\n",
      "Use header DB\tfalse\n",
      "Verbosity    \t3\n",
      "\n",
      "Start writing file to ./fasta/high_resolution_2_protein_sequences_after_2000_representatives.fasta\n",
      "Time for processing: 0h 0m 0s 24ms\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Ensure tmpDir exists\n",
    "os.makedirs('./mmseq_db/tmpDir', exist_ok=True)\n",
    "\n",
    "# Step 1: create sequence DB\n",
    "os.system('mmseqs createdb ./fasta/high_resolution_2_protein_sequences_after_2000.fasta ./mmseq_db/inputDB')\n",
    "\n",
    "# Step 2: cluster sequences\n",
    "os.system('mmseqs cluster ./mmseq_db/inputDB ./mmseq_db/clusterDB ./mmseq_db/tmpDir --min-seq-id 0.3 -c 0.8 --cov-mode 1')\n",
    "\n",
    "# Step 3: extract representative sequences\n",
    "os.system('mmseqs result2repseq ./mmseq_db/inputDB ./mmseq_db/clusterDB ./mmseq_db/repDB')\n",
    "\n",
    "# Step 4: convert the repDB to FASTA\n",
    "os.system('mmseqs convert2fasta ./mmseq_db/repDB ./fasta/high_resolution_2_protein_sequences_after_2000_representatives.fasta')\n",
    "\n",
    "# Clear mmseq_db dir\n",
    "os.system('rm mmseq_db/* -rf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "4a733c58-f0cc-43cf-9440-becb668acec0",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "createdb ./fasta/high_resolution_2_membrane_protein_sequences_after_2000.fasta ./mmseq_db/inputDB \n",
      "\n",
      "MMseqs Version:       \t13.45111\n",
      "Database type         \t0\n",
      "Shuffle input database\ttrue\n",
      "Createdb mode         \t0\n",
      "Write lookup file     \t1\n",
      "Offset of numeric ids \t0\n",
      "Compressed            \t0\n",
      "Verbosity             \t3\n",
      "\n",
      "Converting sequences\n",
      "[\n",
      "Time for merging to inputDB_h: 0h 0m 0s 2ms\n",
      "Time for merging to inputDB: 0h 0m 0s 3ms\n",
      "Database type: Aminoacid\n",
      "Time for processing: 0h 0m 0s 17ms\n",
      "cluster ./mmseq_db/inputDB ./mmseq_db/clusterDB ./mmseq_db/tmpDir --min-seq-id 0.3 -c 0.8 --cov-mode 1 \n",
      "\n",
      "MMseqs Version:                     \t13.45111\n",
      "Substitution matrix                 \tnucl:nucleotide.out,aa:blosum62.out\n",
      "Seed substitution matrix            \tnucl:nucleotide.out,aa:VTML80.out\n",
      "Sensitivity                         \t4\n",
      "k-mer length                        \t0\n",
      "k-score                             \t2147483647\n",
      "Alphabet size                       \tnucl:5,aa:21\n",
      "Max sequence length                 \t65535\n",
      "Max results per query               \t20\n",
      "Split database                      \t0\n",
      "Split mode                          \t2\n",
      "Split memory limit                  \t0\n",
      "Coverage threshold                  \t0.8\n",
      "Coverage mode                       \t1\n",
      "Compositional bias                  \t1\n",
      "Diagonal scoring                    \ttrue\n",
      "Exact k-mer matching                \t0\n",
      "Mask residues                       \t1\n",
      "Mask lower case residues            \t0\n",
      "Minimum diagonal score              \t15\n",
      "Include identical seq. id.          \tfalse\n",
      "Spaced k-mers                       \t1\n",
      "Preload mode                        \t0\n",
      "Pseudo count a                      \t1\n",
      "Pseudo count b                      \t1.5\n",
      "Spaced k-mer pattern                \t\n",
      "Local temporary path                \t\n",
      "Threads                             \t22\n",
      "Compressed                          \t0\n",
      "Verbosity                           \t3\n",
      "Add backtrace                       \tfalse\n",
      "Alignment mode                      \t3\n",
      "Alignment mode                      \t0\n",
      "Allow wrapped scoring               \tfalse\n",
      "E-value threshold                   \t0.001\n",
      "Seq. id. threshold                  \t0.3\n",
      "Min alignment length                \t0\n",
      "Seq. id. mode                       \t0\n",
      "Alternative alignments              \t0\n",
      "Max reject                          \t2147483647\n",
      "Max accept                          \t2147483647\n",
      "Score bias                          \t0\n",
      "Realign hits                        \tfalse\n",
      "Realign score bias                  \t-0.2\n",
      "Realign max seqs                    \t2147483647\n",
      "Gap open cost                       \tnucl:5,aa:11\n",
      "Gap extension cost                  \tnucl:2,aa:1\n",
      "Zdrop                               \t40\n",
      "Rescore mode                        \t0\n",
      "Remove hits by seq. id. and coverage\tfalse\n",
      "Sort results                        \t0\n",
      "Cluster mode                        \t0\n",
      "Max connected component depth       \t1000\n",
      "Similarity type                     \t2\n",
      "Single step clustering              \tfalse\n",
      "Cascaded clustering steps           \t3\n",
      "Cluster reassign                    \tfalse\n",
      "Remove temporary files              \tfalse\n",
      "Force restart with latest tmp       \tfalse\n",
      "MPI runner                          \t\n",
      "k-mers per sequence                 \t21\n",
      "Scale k-mers per sequence           \tnucl:0.200,aa:0.000\n",
      "Adjust k-mer length                 \tfalse\n",
      "Shift hash                          \t67\n",
      "Include only extendable             \tfalse\n",
      "Skip repeating k-mers               \tfalse\n",
      "\n",
      "Set cluster sensitivity to -s 5.000000\n",
      "Set cluster mode GREEDY MEM\n",
      "Set cluster iterations to 3\n",
      "linclust ./mmseq_db/inputDB ./mmseq_db/tmpDir/2469773849256502068/clu_redundancy ./mmseq_db/tmpDir/2469773849256502068/linclust --cluster-mode 3 --max-iterations 1000 --similarity-type 2 --threads 22 --compressed 0 -v 3 --sub-mat nucl:nucleotide.out,aa:blosum62.out -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.3 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 1 --max-seq-len 65535 --comp-bias-corr 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca 1 --pcb 1.5 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --gap-open nucl:5,aa:11 --gap-extend nucl:2,aa:1 --zdrop 40 --alph-size nucl:5,aa:13 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale nucl:0.200,aa:0.000 --adjust-kmer-len 0 --mask 0 --mask-lower-case 0 -k 0 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --rescore-mode 0 --filter-hits 0 --sort-results 0 --remove-tmp-files 0 --force-reuse 0 \n",
      "\n",
      "kmermatcher ./mmseq_db/inputDB ./mmseq_db/tmpDir/2469773849256502068/linclust/13123397006310254668/pref --sub-mat nucl:nucleotide.out,aa:blosum62.out --alph-size nucl:5,aa:13 --min-seq-id 0.3 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale nucl:0.200,aa:0.000 --adjust-kmer-len 0 --mask 0 --mask-lower-case 0 --cov-mode 1 -k 0 -c 0.8 --max-seq-len 65535 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --threads 22 --compressed 0 -v 3 \n",
      "\n",
      "kmermatcher ./mmseq_db/inputDB ./mmseq_db/tmpDir/2469773849256502068/linclust/13123397006310254668/pref --sub-mat nucl:nucleotide.out,aa:blosum62.out --alph-size nucl:5,aa:13 --min-seq-id 0.3 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale nucl:0.200,aa:0.000 --adjust-kmer-len 0 --mask 0 --mask-lower-case 0 --cov-mode 1 -k 0 -c 0.8 --max-seq-len 65535 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --threads 22 --compressed 0 -v 3 \n",
      "\n",
      "Database size: 4606 type: Aminoacid\n",
      "Reduced amino acid alphabet: (A S T) (C) (D B N) (E Q Z) (F Y) (G) (H) (I V) (K R) (L J M) (P) (W) (X) \n",
      "\n",
      "Generate k-mers list for 1 split\n",
      "[=================================================================] 4.61K 0s 12ms\n",
      "Sort kmer 0h 0m 0s 32ms\n",
      "Sort by rep. sequence 0h 0m 0s 34ms\n",
      "Time for fill: 0h 0m 0s 0ms\n",
      "Time for merging to pref: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 116ms\n",
      "rescorediagonal ./mmseq_db/inputDB ./mmseq_db/inputDB ./mmseq_db/tmpDir/2469773849256502068/linclust/13123397006310254668/pref ./mmseq_db/tmpDir/2469773849256502068/linclust/13123397006310254668/pref_rescore1 --sub-mat nucl:nucleotide.out,aa:blosum62.out --rescore-mode 0 --wrapped-scoring 0 --filter-hits 0 -e 0.001 -c 0.8 -a 0 --cov-mode 1 --min-seq-id 0.5 --min-aln-len 0 --seq-id-mode 0 --add-self-matches 0 --sort-results 0 --db-load-mode 0 --threads 22 --compressed 0 -v 3 \n",
      "\n",
      "[=================================================================] 4.61K 0s 3ms\n",
      "Time for merging to pref_rescore1: 0h 0m 0s 1ms\n",
      "Time for processing: 0h 0m 0s 16ms\n",
      "clust ./mmseq_db/inputDB ./mmseq_db/tmpDir/2469773849256502068/linclust/13123397006310254668/pref_rescore1 ./mmseq_db/tmpDir/2469773849256502068/linclust/13123397006310254668/pre_clust --cluster-mode 3 --max-iterations 1000 --similarity-type 2 --threads 22 --compressed 0 -v 3 \n",
      "\n",
      "Clustering mode: Greedy Low Mem\n",
      "Total time: 0h 0m 0s 32ms\n",
      "\n",
      "Size of the sequence database: 4606\n",
      "Size of the alignment database: 4606\n",
      "Number of clusters: 881\n",
      "\n",
      "Writing results 0h 0m 0s 0ms\n",
      "Time for merging to pre_clust: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 34ms\n",
      "createsubdb ./mmseq_db/tmpDir/2469773849256502068/linclust/13123397006310254668/order_redundancy ./mmseq_db/inputDB ./mmseq_db/tmpDir/2469773849256502068/linclust/13123397006310254668/input_step_redundancy -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step_redundancy: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "createsubdb ./mmseq_db/tmpDir/2469773849256502068/linclust/13123397006310254668/order_redundancy ./mmseq_db/tmpDir/2469773849256502068/linclust/13123397006310254668/pref ./mmseq_db/tmpDir/2469773849256502068/linclust/13123397006310254668/pref_filter1 -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to pref_filter1: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 1ms\n",
      "filterdb ./mmseq_db/tmpDir/2469773849256502068/linclust/13123397006310254668/pref_filter1 ./mmseq_db/tmpDir/2469773849256502068/linclust/13123397006310254668/pref_filter2 --filter-file ./mmseq_db/tmpDir/2469773849256502068/linclust/13123397006310254668/order_redundancy --threads 22 --compressed 0 -v 3 \n",
      "\n",
      "Filtering using file(s)\n",
      "[=================================================================] 881 0s 4ms\n",
      "Time for merging to pref_filter2: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 17ms\n",
      "rescorediagonal ./mmseq_db/tmpDir/2469773849256502068/linclust/13123397006310254668/input_step_redundancy ./mmseq_db/tmpDir/2469773849256502068/linclust/13123397006310254668/input_step_redundancy ./mmseq_db/tmpDir/2469773849256502068/linclust/13123397006310254668/pref_filter2 ./mmseq_db/tmpDir/2469773849256502068/linclust/13123397006310254668/pref_rescore2 --sub-mat nucl:nucleotide.out,aa:blosum62.out --rescore-mode 1 --wrapped-scoring 0 --filter-hits 1 -e 0.001 -c 0.8 -a 0 --cov-mode 1 --min-seq-id 0.3 --min-aln-len 0 --seq-id-mode 0 --add-self-matches 0 --sort-results 0 --db-load-mode 0 --threads 22 --compressed 0 -v 3 \n",
      "\n",
      "[=================================================================] 881 0s 6ms\n",
      "Time for merging to pref_rescore2: 0h 0m 0s 1ms\n",
      "Time for processing: 0h 0m 0s 16ms\n",
      "align ./mmseq_db/tmpDir/2469773849256502068/linclust/13123397006310254668/input_step_redundancy ./mmseq_db/tmpDir/2469773849256502068/linclust/13123397006310254668/input_step_redundancy ./mmseq_db/tmpDir/2469773849256502068/linclust/13123397006310254668/pref_rescore2 ./mmseq_db/tmpDir/2469773849256502068/linclust/13123397006310254668/aln --sub-mat nucl:nucleotide.out,aa:blosum62.out -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.3 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 1 --max-seq-len 65535 --comp-bias-corr 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca 1 --pcb 1.5 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --gap-open nucl:5,aa:11 --gap-extend nucl:2,aa:1 --zdrop 40 --threads 22 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 881 type: Aminoacid\n",
      "Target database size: 881 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 881 0s 15ms\n",
      "Time for merging to aln: 0h 0m 0s 4ms\n",
      "924 alignments calculated\n",
      "910 sequence pairs passed the thresholds (0.984848 of overall calculated)\n",
      "1.032917 hits per query sequence\n",
      "Time for processing: 0h 0m 0s 26ms\n",
      "clust ./mmseq_db/tmpDir/2469773849256502068/linclust/13123397006310254668/input_step_redundancy ./mmseq_db/tmpDir/2469773849256502068/linclust/13123397006310254668/aln ./mmseq_db/tmpDir/2469773849256502068/linclust/13123397006310254668/clust --cluster-mode 3 --max-iterations 1000 --similarity-type 2 --threads 22 --compressed 0 -v 3 \n",
      "\n",
      "Clustering mode: Greedy Low Mem\n",
      "Total time: 0h 0m 0s 47ms\n",
      "\n",
      "Size of the sequence database: 881\n",
      "Size of the alignment database: 881\n",
      "Number of clusters: 832\n",
      "\n",
      "Writing results 0h 0m 0s 0ms\n",
      "Time for merging to clust: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 48ms\n",
      "mergeclusters ./mmseq_db/inputDB ./mmseq_db/tmpDir/2469773849256502068/clu_redundancy ./mmseq_db/tmpDir/2469773849256502068/linclust/13123397006310254668/pre_clust ./mmseq_db/tmpDir/2469773849256502068/linclust/13123397006310254668/clust --threads 22 --compressed 0 -v 3 \n",
      "\n",
      "Clustering step 1\n",
      "[=================================================================] 881 0s 5ms\n",
      "Clustering step 2\n",
      "[=================================================================] 832 0s 21ms\n",
      "Write merged clustering\n",
      "[=================================================================] 4.61K 0s 30ms\n",
      "Time for merging to clu_redundancy: 0h 0m 0s 10ms\n",
      "Time for processing: 0h 0m 0s 55ms\n",
      "createsubdb ./mmseq_db/tmpDir/2469773849256502068/clu_redundancy ./mmseq_db/inputDB ./mmseq_db/tmpDir/2469773849256502068/input_step_redundancy -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step_redundancy: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "prefilter ./mmseq_db/tmpDir/2469773849256502068/input_step_redundancy ./mmseq_db/tmpDir/2469773849256502068/input_step_redundancy ./mmseq_db/tmpDir/2469773849256502068/pref_step0 --sub-mat nucl:nucleotide.out,aa:blosum62.out --seed-sub-mat nucl:nucleotide.out,aa:VTML80.out -s 1 -k 0 --k-score 2147483647 --alph-size nucl:5,aa:21 --max-seq-len 65535 --max-seqs 20 --split 0 --split-mode 2 --split-memory-limit 0 -c 0.8 --cov-mode 1 --comp-bias-corr 0 --diag-score 0 --exact-kmer-matching 0 --mask 1 --mask-lower-case 0 --min-ungapped-score 0 --add-self-matches 0 --spaced-kmer-mode 1 --db-load-mode 0 --pca 1 --pcb 1.5 --threads 22 --compressed 0 -v 3 \n",
      "\n",
      "Query database size: 832 type: Aminoacid\n",
      "Estimated memory consumption: 980M\n",
      "Target database size: 832 type: Aminoacid\n",
      "Index table k-mer threshold: 154 at k-mer size 6 \n",
      "Index table: counting k-mers\n",
      "[=================================================================] 832 0s 18ms\n",
      "Index table: Masked residues: 2079\n",
      "Index table: fill\n",
      "[=================================================================] 832 0s 7ms\n",
      "Index statistics\n",
      "Entries:          97433\n",
      "DB size:          488 MB\n",
      "Avg k-mer size:   0.001522\n",
      "Top 10 k-mers\n",
      "    SGVRHM\t20\n",
      "    RLMAGF\t7\n",
      "    SGHNVT\t7\n",
      "    GPPPGN\t5\n",
      "    RYDVTP\t5\n",
      "    DLTNFQ\t5\n",
      "    YWQTTV\t5\n",
      "    LKETYY\t5\n",
      "    ALYFVF\t4\n",
      "    DITEYF\t4\n",
      "Time for index table init: 0h 0m 0s 235ms\n",
      "Process prefiltering step 1 of 1\n",
      "\n",
      "k-mer similarity threshold: 154\n",
      "Starting prefiltering scores calculation (step 1 of 1)\n",
      "Query db start 1 to 832\n",
      "Target db start 1 to 832\n",
      "[=================================================================] 832 0s 8ms\n",
      "\n",
      "-nan k-mers per position\n",
      "122 DB matches per sequence\n",
      "0 overflows\n",
      "0 queries produce too many hits (truncated result)\n",
      "1 sequences passed prefiltering per query sequence\n",
      "1 median result list length\n",
      "0 sequences with 0 size result lists\n",
      "Time for merging to pref_step0: 0h 0m 0s 1ms\n",
      "Time for processing: 0h 0m 0s 467ms\n",
      "align ./mmseq_db/tmpDir/2469773849256502068/input_step_redundancy ./mmseq_db/tmpDir/2469773849256502068/input_step_redundancy ./mmseq_db/tmpDir/2469773849256502068/pref_step0 ./mmseq_db/tmpDir/2469773849256502068/aln_step0 --sub-mat nucl:nucleotide.out,aa:blosum62.out -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.3 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 1 --max-seq-len 65535 --comp-bias-corr 0 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca 1 --pcb 1.5 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --gap-open nucl:5,aa:11 --gap-extend nucl:2,aa:1 --zdrop 40 --threads 22 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 832 type: Aminoacid\n",
      "Target database size: 832 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 832 0s 22ms\n",
      "Time for merging to aln_step0: 0h 0m 0s 1ms\n",
      "1397 alignments calculated\n",
      "1067 sequence pairs passed the thresholds (0.763780 of overall calculated)\n",
      "1.282452 hits per query sequence\n",
      "Time for processing: 0h 0m 0s 28ms\n",
      "clust ./mmseq_db/tmpDir/2469773849256502068/input_step_redundancy ./mmseq_db/tmpDir/2469773849256502068/aln_step0 ./mmseq_db/tmpDir/2469773849256502068/clu_step0 --cluster-mode 3 --max-iterations 1000 --similarity-type 2 --threads 22 --compressed 0 -v 3 \n",
      "\n",
      "Clustering mode: Greedy Low Mem\n",
      "Total time: 0h 0m 0s 16ms\n",
      "\n",
      "Size of the sequence database: 832\n",
      "Size of the alignment database: 832\n",
      "Number of clusters: 741\n",
      "\n",
      "Writing results 0h 0m 0s 0ms\n",
      "Time for merging to clu_step0: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 18ms\n",
      "createsubdb ./mmseq_db/tmpDir/2469773849256502068/clu_step0 ./mmseq_db/tmpDir/2469773849256502068/input_step_redundancy ./mmseq_db/tmpDir/2469773849256502068/input_step1 -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step1: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "prefilter ./mmseq_db/tmpDir/2469773849256502068/input_step1 ./mmseq_db/tmpDir/2469773849256502068/input_step1 ./mmseq_db/tmpDir/2469773849256502068/pref_step1 --sub-mat nucl:nucleotide.out,aa:blosum62.out --seed-sub-mat nucl:nucleotide.out,aa:VTML80.out -s 3 -k 0 --k-score 2147483647 --alph-size nucl:5,aa:21 --max-seq-len 65535 --max-seqs 20 --split 0 --split-mode 2 --split-memory-limit 0 -c 0.8 --cov-mode 1 --comp-bias-corr 1 --diag-score 1 --exact-kmer-matching 0 --mask 1 --mask-lower-case 0 --min-ungapped-score 15 --add-self-matches 0 --spaced-kmer-mode 1 --db-load-mode 0 --pca 1 --pcb 1.5 --threads 22 --compressed 0 -v 3 \n",
      "\n",
      "Query database size: 741 type: Aminoacid\n",
      "Estimated memory consumption: 979M\n",
      "Target database size: 741 type: Aminoacid\n",
      "Index table k-mer threshold: 136 at k-mer size 6 \n",
      "Index table: counting k-mers\n",
      "[=================================================================] 741 0s 9ms\n",
      "Index table: Masked residues: 1969\n",
      "Index table: fill\n",
      "[=================================================================] 741 0s 17ms\n",
      "Index statistics\n",
      "Entries:          161168\n",
      "DB size:          489 MB\n",
      "Avg k-mer size:   0.002518\n",
      "Top 10 k-mers\n",
      "    SSLPSH\t19\n",
      "    SGVRHM\t19\n",
      "    SGDGNL\t5\n",
      "    DLTNFQ\t5\n",
      "    LGELQS\t5\n",
      "    GVLTLY\t5\n",
      "    FCVAKD\t4\n",
      "    KTILIF\t4\n",
      "    ALYFVF\t4\n",
      "    DITEYF\t4\n",
      "Time for index table init: 0h 0m 0s 278ms\n",
      "Process prefiltering step 1 of 1\n",
      "\n",
      "k-mer similarity threshold: 136\n",
      "Starting prefiltering scores calculation (step 1 of 1)\n",
      "Query db start 1 to 741\n",
      "Target db start 1 to 741\n",
      "[=================================================================] 741 0s 29ms\n",
      "\n",
      "-nan k-mers per position\n",
      "217 DB matches per sequence\n",
      "0 overflows\n",
      "0 queries produce too many hits (truncated result)\n",
      "1 sequences passed prefiltering per query sequence\n",
      "1 median result list length\n",
      "0 sequences with 0 size result lists\n",
      "Time for merging to pref_step1: 0h 0m 0s 1ms\n",
      "Time for processing: 0h 0m 0s 571ms\n",
      "align ./mmseq_db/tmpDir/2469773849256502068/input_step1 ./mmseq_db/tmpDir/2469773849256502068/input_step1 ./mmseq_db/tmpDir/2469773849256502068/pref_step1 ./mmseq_db/tmpDir/2469773849256502068/aln_step1 --sub-mat nucl:nucleotide.out,aa:blosum62.out -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.3 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 1 --max-seq-len 65535 --comp-bias-corr 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca 1 --pcb 1.5 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --gap-open nucl:5,aa:11 --gap-extend nucl:2,aa:1 --zdrop 40 --threads 22 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 741 type: Aminoacid\n",
      "Target database size: 741 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 741 0s 16ms\n",
      "Time for merging to aln_step1: 0h 0m 0s 1ms\n",
      "1173 alignments calculated\n",
      "778 sequence pairs passed the thresholds (0.663257 of overall calculated)\n",
      "1.049932 hits per query sequence\n",
      "Time for processing: 0h 0m 0s 22ms\n",
      "clust ./mmseq_db/tmpDir/2469773849256502068/input_step1 ./mmseq_db/tmpDir/2469773849256502068/aln_step1 ./mmseq_db/tmpDir/2469773849256502068/clu_step1 --cluster-mode 3 --max-iterations 1000 --similarity-type 2 --threads 22 --compressed 0 -v 3 \n",
      "\n",
      "Clustering mode: Greedy Low Mem\n",
      "Total time: 0h 0m 0s 16ms\n",
      "\n",
      "Size of the sequence database: 741\n",
      "Size of the alignment database: 741\n",
      "Number of clusters: 714\n",
      "\n",
      "Writing results 0h 0m 0s 0ms\n",
      "Time for merging to clu_step1: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 17ms\n",
      "createsubdb ./mmseq_db/tmpDir/2469773849256502068/clu_step1 ./mmseq_db/tmpDir/2469773849256502068/input_step1 ./mmseq_db/tmpDir/2469773849256502068/input_step2 -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step2: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "prefilter ./mmseq_db/tmpDir/2469773849256502068/input_step2 ./mmseq_db/tmpDir/2469773849256502068/input_step2 ./mmseq_db/tmpDir/2469773849256502068/pref_step2 --sub-mat nucl:nucleotide.out,aa:blosum62.out --seed-sub-mat nucl:nucleotide.out,aa:VTML80.out -s 5 -k 0 --k-score 2147483647 --alph-size nucl:5,aa:21 --max-seq-len 65535 --max-seqs 20 --split 0 --split-mode 2 --split-memory-limit 0 -c 0.8 --cov-mode 1 --comp-bias-corr 1 --diag-score 1 --exact-kmer-matching 0 --mask 1 --mask-lower-case 0 --min-ungapped-score 15 --add-self-matches 0 --spaced-kmer-mode 1 --db-load-mode 0 --pca 1 --pcb 1.5 --threads 22 --compressed 0 -v 3 \n",
      "\n",
      "Query database size: 714 type: Aminoacid\n",
      "Estimated memory consumption: 979M\n",
      "Target database size: 714 type: Aminoacid\n",
      "Index table k-mer threshold: 118 at k-mer size 6 \n",
      "Index table: counting k-mers\n",
      "[=================================================================] 714 0s 16ms\n",
      "Index table: Masked residues: 1932\n",
      "Index table: fill\n",
      "[=================================================================] 714 0s 15ms\n",
      "Index statistics\n",
      "Entries:          161648\n",
      "DB size:          489 MB\n",
      "Avg k-mer size:   0.002526\n",
      "Top 10 k-mers\n",
      "    SSLPSH\t18\n",
      "    SGVRHM\t18\n",
      "    SGDGNL\t5\n",
      "    SSVLEN\t5\n",
      "    DLTNFQ\t5\n",
      "    LGELQS\t5\n",
      "    GVLTLY\t5\n",
      "    KTILIF\t4\n",
      "    ALYFVF\t4\n",
      "    DITEYF\t4\n",
      "Time for index table init: 0h 0m 0s 274ms\n",
      "Process prefiltering step 1 of 1\n",
      "\n",
      "k-mer similarity threshold: 118\n",
      "Starting prefiltering scores calculation (step 1 of 1)\n",
      "Query db start 1 to 714\n",
      "Target db start 1 to 714\n",
      "[=================================================================] 714 0s 81ms\n",
      "\n",
      "-nan k-mers per position\n",
      "327 DB matches per sequence\n",
      "0 overflows\n",
      "0 queries produce too many hits (truncated result)\n",
      "2 sequences passed prefiltering per query sequence\n",
      "2 median result list length\n",
      "0 sequences with 0 size result lists\n",
      "Time for merging to pref_step2: 0h 0m 0s 1ms\n",
      "Time for processing: 0h 0m 0s 647ms\n",
      "align ./mmseq_db/tmpDir/2469773849256502068/input_step2 ./mmseq_db/tmpDir/2469773849256502068/input_step2 ./mmseq_db/tmpDir/2469773849256502068/pref_step2 ./mmseq_db/tmpDir/2469773849256502068/aln_step2 --sub-mat nucl:nucleotide.out,aa:blosum62.out -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.3 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 1 --max-seq-len 65535 --comp-bias-corr 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca 1 --pcb 1.5 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --gap-open nucl:5,aa:11 --gap-extend nucl:2,aa:1 --zdrop 40 --threads 22 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 714 type: Aminoacid\n",
      "Target database size: 714 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 714 0s 19ms\n",
      "Time for merging to aln_step2: 0h 0m 0s 2ms\n",
      "1513 alignments calculated\n",
      "731 sequence pairs passed the thresholds (0.483146 of overall calculated)\n",
      "1.023810 hits per query sequence\n",
      "Time for processing: 0h 0m 0s 29ms\n",
      "clust ./mmseq_db/tmpDir/2469773849256502068/input_step2 ./mmseq_db/tmpDir/2469773849256502068/aln_step2 ./mmseq_db/tmpDir/2469773849256502068/clu_step2 --cluster-mode 3 --max-iterations 1000 --similarity-type 2 --threads 22 --compressed 0 -v 3 \n",
      "\n",
      "Clustering mode: Greedy Low Mem\n",
      "Total time: 0h 0m 0s 20ms\n",
      "\n",
      "Size of the sequence database: 714\n",
      "Size of the alignment database: 714\n",
      "Number of clusters: 688\n",
      "\n",
      "Writing results 0h 0m 0s 0ms\n",
      "Time for merging to clu_step2: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 21ms\n",
      "mergeclusters ./mmseq_db/inputDB ./mmseq_db/clusterDB ./mmseq_db/tmpDir/2469773849256502068/clu_redundancy ./mmseq_db/tmpDir/2469773849256502068/clu_step0 ./mmseq_db/tmpDir/2469773849256502068/clu_step1 ./mmseq_db/tmpDir/2469773849256502068/clu_step2 --threads 22 --compressed 0 -v 3 \n",
      "\n",
      "Clustering step 1\n",
      "[=================================================================] 832 0s 7ms\n",
      "Clustering step 2\n",
      "[=================================================================] 741 0s 18ms\n",
      "Clustering step 3\n",
      "[=================================================================] 714 0s 30ms\n",
      "Clustering step 4\n",
      "[=================================================================] 688 0s 43ms\n",
      "Write merged clustering\n",
      "[=================================================================] 4.61K 0s 49ms\n",
      "Time for merging to clusterDB: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 60ms\n",
      "result2repseq ./mmseq_db/inputDB ./mmseq_db/clusterDB ./mmseq_db/repDB \n",
      "\n",
      "MMseqs Version:\t13.45111\n",
      "Preload mode\t0\n",
      "Compressed  \t0\n",
      "Threads     \t22\n",
      "Verbosity   \t3\n",
      "\n",
      "[=================================================================] 688 0s 7ms\n",
      "Time for merging to repDB: 0h 0m 0s 6ms\n",
      "Time for processing: 0h 0m 0s 21ms\n",
      "convert2fasta ./mmseq_db/repDB ./fasta/high_resolution_2_membrane_protein_sequences_after_2000_representatives.fasta \n",
      "\n",
      "MMseqs Version:\t13.45111\n",
      "Use header DB\tfalse\n",
      "Verbosity    \t3\n",
      "\n",
      "Start writing file to ./fasta/high_resolution_2_membrane_protein_sequences_after_2000_representatives.fasta\n",
      "Time for processing: 0h 0m 0s 0ms\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Ensure tmpDir exists\n",
    "os.makedirs('./mmseq_db/tmpDir', exist_ok=True)\n",
    "\n",
    "# Step 1: create sequence DB\n",
    "os.system('mmseqs createdb ./fasta/high_resolution_2_membrane_protein_sequences_after_2000.fasta ./mmseq_db/inputDB')\n",
    "\n",
    "# Step 2: cluster sequences\n",
    "os.system('mmseqs cluster ./mmseq_db/inputDB ./mmseq_db/clusterDB ./mmseq_db/tmpDir --min-seq-id 0.3 -c 0.8 --cov-mode 1')\n",
    "\n",
    "# Step 3: extract representative sequences\n",
    "os.system('mmseqs result2repseq ./mmseq_db/inputDB ./mmseq_db/clusterDB ./mmseq_db/repDB')\n",
    "\n",
    "# Step 4: convert the repDB to FASTA\n",
    "os.system('mmseqs convert2fasta ./mmseq_db/repDB ./fasta/high_resolution_2_membrane_protein_sequences_after_2000_representatives.fasta')\n",
    "\n",
    "# Clear mmseq_db dir\n",
    "os.system('rm mmseq_db/* -rf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "7f87a559-b4e2-4a99-8d4e-3a3bdbc3160e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "688 565\n"
     ]
    }
   ],
   "source": [
    "with open('./fasta/high_resolution_2_membrane_protein_sequences_after_2000_representatives.fasta', 'r') as f:\n",
    "    pdb_chain = []\n",
    "    for line in f.readlines():\n",
    "        if line[0]=='>':\n",
    "            pdb_chain.append(line[1:])\n",
    "pdb_chain_arr = np.array([i.split('_') for i in pdb_chain])\n",
    "print(len(pdb_chain_arr), len(np.unique(pdb_chain_arr[:,0])))\n",
    "np.save('mem_prot_rep.npy', pdb_chain_arr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "a2d62ae0-6cdb-4415-b219-edcbd9535f08",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "17704 16683\n"
     ]
    }
   ],
   "source": [
    "with open('./fasta/high_resolution_2_protein_sequences_after_2000_representatives.fasta', 'r') as f:\n",
    "    pdb_chain = []\n",
    "    for line in f.readlines():\n",
    "        if line[0]=='>':\n",
    "            pdb_chain.append(line[1:])\n",
    "pdb_chain_arr = np.array([i.split('_') for i in pdb_chain])\n",
    "print(len(pdb_chain_arr), len(np.unique(pdb_chain_arr[:,0])))\n",
    "np.save('prot_rep.npy', pdb_chain_arr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "394b0ffa-aad3-4cc0-a60a-6897b42950ff",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.23"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
