{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "120ef59f-945a-4fe5-9337-994e062657ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import csv\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "c9f8578c-6ee9-480e-bd17-28f57b3cbc99",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "###\n",
    "#LEGACY\n",
    "###\n",
    "\n",
    "# with open('./data/high_resolution_2_membrane_protein_sequences_after_2000.csv', newline='') as f:\n",
    "#     reader = csv.reader(f)\n",
    "#     file = list(reader)  # convert the reader to a list of rows\n",
    "# membrane_pdb_ids = []\n",
    "# for line in file[2:]:\n",
    "#     if len(line[0])==4:\n",
    "#         pdb_id = line[0]\n",
    "#         membrane_pdb_ids.append(pdb_id)\n",
    "#     elif len(line[0])>1:\n",
    "#         pdb_id = line[0][:2]+line[0][-2:]\n",
    "#         membrane_pdb_ids.append(pdb_id)\n",
    "        \n",
    "# with open('./data/high_resolution_2_protein_sequences_after_2000.csv', newline='') as f:\n",
    "#     reader = csv.reader(f)\n",
    "#     file = list(reader)  # convert the reader to a list of rows\n",
    "\n",
    "# fasta = []\n",
    "# names = []\n",
    "# for line in file[2:]:\n",
    "#     if len(line[0])==4:\n",
    "#         pdb_id = line[0]\n",
    "#     elif len(line[0])>1:\n",
    "#         pdb_id = line[0][:2]+line[0][-2:]\n",
    "#     if not len(line[2]) or line[0] == 'Entry ID' or line[-2] != 'polypeptide(L)' or pdb_id in membrane_pdb_ids:\n",
    "#         continue\n",
    "#     #line = line.split(',')\n",
    "#     chains = [i.strip() for i in line[2].split(',')]\n",
    "#     if len(chains)-1:\n",
    "#         for c in chains:\n",
    "#             fasta.append(f'>{pdb_id}_{c}')\n",
    "#             names.append(f'>{pdb_id}_{c}')\n",
    "#             fasta.append(line[6])\n",
    "#     else:\n",
    "#         fasta.append(f'>{pdb_id}_{chains[0]}')\n",
    "#         names.append(f'>{pdb_id}_{chains[0]}')\n",
    "#         fasta.append(line[6])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "38af7f81-d4ab-45d8-8cb5-6dba3e7f478e",
   "metadata": {},
   "outputs": [],
   "source": [
    "path_mem_proteins = './data/high_resolution_2dot8_membrane_protein_sequences_after_2000.csv'\n",
    "with open(path_mem_proteins, newline='') as f:\n",
    "    reader = csv.reader(f)\n",
    "    mem_file = np.array(list(reader))  # convert the reader to a list of rows\n",
    "membrane_pdb_ids = mem_file[:,0]\n",
    "\n",
    "path_all_proteins = './data/high_resolution_2dot8_protein_sequences_after_2000.csv'\n",
    "with open(path_all_proteins, newline='') as f:\n",
    "    reader = csv.reader(f)\n",
    "    prot_file = np.array(list(reader))  # convert the reader to a list of rows"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "846cfcd3-fe11-41ba-9948-0a4d533746f4",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'prot_file' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[28], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m file \u001b[38;5;241m=\u001b[39m np\u001b[38;5;241m.\u001b[39mcopy(\u001b[43mprot_file\u001b[49m)\n\u001b[1;32m      2\u001b[0m fasta \u001b[38;5;241m=\u001b[39m []\n\u001b[1;32m      3\u001b[0m names \u001b[38;5;241m=\u001b[39m []\n",
      "\u001b[0;31mNameError\u001b[0m: name 'prot_file' is not defined"
     ]
    }
   ],
   "source": [
    "file = np.copy(prot_file)\n",
    "fasta = []\n",
    "names = []\n",
    "for line in file:\n",
    "    pdb_id = line[0]\n",
    "    if line[-1] != 'polypeptide(L)' or pdb_id in membrane_pdb_ids:\n",
    "        continue\n",
    "    #line = line.split(',')\n",
    "    chains = [i.strip() for i in line[1].split(',')]\n",
    "    if len(chains)-1:\n",
    "        for c in chains:\n",
    "            fasta.append(f'>{pdb_id}_{c}')\n",
    "            names.append(f'>{pdb_id}_{c}')\n",
    "            fasta.append(line[2])\n",
    "    else:\n",
    "        fasta.append(f'>{pdb_id}_{chains[0]}')\n",
    "        names.append(f'>{pdb_id}_{chains[0]}')\n",
    "        fasta.append(line[2])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "20fbfb0e-4238-437f-977c-9d453585c9eb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "11202 11202\n"
     ]
    }
   ],
   "source": [
    "un, ind, counts = np.unique(names, return_counts=True, return_index=True)\n",
    "print(len(names), len(un))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "63429e97-30da-4ddb-bc5f-2b09366719b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('./fasta/high_resolution_2dot8_membrane_protein_sequences_after_2000.fasta', 'w') as f:\n",
    "    for line in fasta:\n",
    "        f.write(f'{line}\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "6983d69f-279e-4669-9e49-f81978f97da9",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "createdb ./fasta/high_resolution_2dot8_membrane_protein_sequences_after_2000.fasta ./mmseq_db/inputDB \n",
      "\n",
      "MMseqs Version:       \t13.45111\n",
      "Database type         \t0\n",
      "Shuffle input database\ttrue\n",
      "Createdb mode         \t0\n",
      "Write lookup file     \t1\n",
      "Offset of numeric ids \t0\n",
      "Compressed            \t0\n",
      "Verbosity             \t3\n",
      "\n",
      "Converting sequences\n",
      "[=\n",
      "Time for merging to inputDB_h: 0h 0m 0s 1ms\n",
      "Time for merging to inputDB: 0h 0m 0s 2ms\n",
      "Database type: Aminoacid\n",
      "Time for processing: 0h 0m 0s 18ms\n",
      "cluster ./mmseq_db/inputDB ./mmseq_db/clusterDB ./mmseq_db/tmpDir --min-seq-id 0.3 -c 0.8 --cov-mode 1 \n",
      "\n",
      "MMseqs Version:                     \t13.45111\n",
      "Substitution matrix                 \tnucl:nucleotide.out,aa:blosum62.out\n",
      "Seed substitution matrix            \tnucl:nucleotide.out,aa:VTML80.out\n",
      "Sensitivity                         \t4\n",
      "k-mer length                        \t0\n",
      "k-score                             \t2147483647\n",
      "Alphabet size                       \tnucl:5,aa:21\n",
      "Max sequence length                 \t65535\n",
      "Max results per query               \t20\n",
      "Split database                      \t0\n",
      "Split mode                          \t2\n",
      "Split memory limit                  \t0\n",
      "Coverage threshold                  \t0.8\n",
      "Coverage mode                       \t1\n",
      "Compositional bias                  \t1\n",
      "Diagonal scoring                    \ttrue\n",
      "Exact k-mer matching                \t0\n",
      "Mask residues                       \t1\n",
      "Mask lower case residues            \t0\n",
      "Minimum diagonal score              \t15\n",
      "Include identical seq. id.          \tfalse\n",
      "Spaced k-mers                       \t1\n",
      "Preload mode                        \t0\n",
      "Pseudo count a                      \t1\n",
      "Pseudo count b                      \t1.5\n",
      "Spaced k-mer pattern                \t\n",
      "Local temporary path                \t\n",
      "Threads                             \t22\n",
      "Compressed                          \t0\n",
      "Verbosity                           \t3\n",
      "Add backtrace                       \tfalse\n",
      "Alignment mode                      \t3\n",
      "Alignment mode                      \t0\n",
      "Allow wrapped scoring               \tfalse\n",
      "E-value threshold                   \t0.001\n",
      "Seq. id. threshold                  \t0.3\n",
      "Min alignment length                \t0\n",
      "Seq. id. mode                       \t0\n",
      "Alternative alignments              \t0\n",
      "Max reject                          \t2147483647\n",
      "Max accept                          \t2147483647\n",
      "Score bias                          \t0\n",
      "Realign hits                        \tfalse\n",
      "Realign score bias                  \t-0.2\n",
      "Realign max seqs                    \t2147483647\n",
      "Gap open cost                       \tnucl:5,aa:11\n",
      "Gap extension cost                  \tnucl:2,aa:1\n",
      "Zdrop                               \t40\n",
      "Rescore mode                        \t0\n",
      "Remove hits by seq. id. and coverage\tfalse\n",
      "Sort results                        \t0\n",
      "Cluster mode                        \t0\n",
      "Max connected component depth       \t1000\n",
      "Similarity type                     \t2\n",
      "Single step clustering              \tfalse\n",
      "Cascaded clustering steps           \t3\n",
      "Cluster reassign                    \tfalse\n",
      "Remove temporary files              \tfalse\n",
      "Force restart with latest tmp       \tfalse\n",
      "MPI runner                          \t\n",
      "k-mers per sequence                 \t21\n",
      "Scale k-mers per sequence           \tnucl:0.200,aa:0.000\n",
      "Adjust k-mer length                 \tfalse\n",
      "Shift hash                          \t67\n",
      "Include only extendable             \tfalse\n",
      "Skip repeating k-mers               \tfalse\n",
      "\n",
      "Set cluster sensitivity to -s 5.000000\n",
      "Set cluster mode GREEDY MEM\n",
      "Set cluster iterations to 3\n",
      "linclust ./mmseq_db/inputDB ./mmseq_db/tmpDir/185048956948561117/clu_redundancy ./mmseq_db/tmpDir/185048956948561117/linclust --cluster-mode 3 --max-iterations 1000 --similarity-type 2 --threads 22 --compressed 0 -v 3 --sub-mat nucl:nucleotide.out,aa:blosum62.out -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.3 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 1 --max-seq-len 65535 --comp-bias-corr 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca 1 --pcb 1.5 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --gap-open nucl:5,aa:11 --gap-extend nucl:2,aa:1 --zdrop 40 --alph-size nucl:5,aa:13 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale nucl:0.200,aa:0.000 --adjust-kmer-len 0 --mask 0 --mask-lower-case 0 -k 0 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --rescore-mode 0 --filter-hits 0 --sort-results 0 --remove-tmp-files 0 --force-reuse 0 \n",
      "\n",
      "kmermatcher ./mmseq_db/inputDB ./mmseq_db/tmpDir/185048956948561117/linclust/3316508196410862791/pref --sub-mat nucl:nucleotide.out,aa:blosum62.out --alph-size nucl:5,aa:13 --min-seq-id 0.3 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale nucl:0.200,aa:0.000 --adjust-kmer-len 0 --mask 0 --mask-lower-case 0 --cov-mode 1 -k 0 -c 0.8 --max-seq-len 65535 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --threads 22 --compressed 0 -v 3 \n",
      "\n",
      "kmermatcher ./mmseq_db/inputDB ./mmseq_db/tmpDir/185048956948561117/linclust/3316508196410862791/pref --sub-mat nucl:nucleotide.out,aa:blosum62.out --alph-size nucl:5,aa:13 --min-seq-id 0.3 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale nucl:0.200,aa:0.000 --adjust-kmer-len 0 --mask 0 --mask-lower-case 0 --cov-mode 1 -k 0 -c 0.8 --max-seq-len 65535 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --threads 22 --compressed 0 -v 3 \n",
      "\n",
      "Database size: 11202 type: Aminoacid\n",
      "Reduced amino acid alphabet: (A S T) (C) (D B N) (E Q Z) (F Y) (G) (H) (I V) (K R) (L J M) (P) (W) (X) \n",
      "\n",
      "Generate k-mers list for 1 split\n",
      "[=================================================================] 11.20K 0s 19ms\n",
      "Sort kmer 0h 0m 0s 51ms\n",
      "Sort by rep. sequence 0h 0m 0s 27ms\n",
      "Time for fill: 0h 0m 0s 5ms\n",
      "Time for merging to pref: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 160ms\n",
      "rescorediagonal ./mmseq_db/inputDB ./mmseq_db/inputDB ./mmseq_db/tmpDir/185048956948561117/linclust/3316508196410862791/pref ./mmseq_db/tmpDir/185048956948561117/linclust/3316508196410862791/pref_rescore1 --sub-mat nucl:nucleotide.out,aa:blosum62.out --rescore-mode 0 --wrapped-scoring 0 --filter-hits 0 -e 0.001 -c 0.8 -a 0 --cov-mode 1 --min-seq-id 0.5 --min-aln-len 0 --seq-id-mode 0 --add-self-matches 0 --sort-results 0 --db-load-mode 0 --threads 22 --compressed 0 -v 3 \n",
      "\n",
      "[=================================================================] 11.20K 0s 8ms\n",
      "Time for merging to pref_rescore1: 0h 0m 0s 10ms\n",
      "Time for processing: 0h 0m 0s 32ms\n",
      "clust ./mmseq_db/inputDB ./mmseq_db/tmpDir/185048956948561117/linclust/3316508196410862791/pref_rescore1 ./mmseq_db/tmpDir/185048956948561117/linclust/3316508196410862791/pre_clust --cluster-mode 3 --max-iterations 1000 --similarity-type 2 --threads 22 --compressed 0 -v 3 \n",
      "\n",
      "Clustering mode: Greedy Low Mem\n",
      "Total time: 0h 0m 0s 36ms\n",
      "\n",
      "Size of the sequence database: 11202\n",
      "Size of the alignment database: 11202\n",
      "Number of clusters: 1737\n",
      "\n",
      "Writing results 0h 0m 0s 0ms\n",
      "Time for merging to pre_clust: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 39ms\n",
      "createsubdb ./mmseq_db/tmpDir/185048956948561117/linclust/3316508196410862791/order_redundancy ./mmseq_db/inputDB ./mmseq_db/tmpDir/185048956948561117/linclust/3316508196410862791/input_step_redundancy -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step_redundancy: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 1ms\n",
      "createsubdb ./mmseq_db/tmpDir/185048956948561117/linclust/3316508196410862791/order_redundancy ./mmseq_db/tmpDir/185048956948561117/linclust/3316508196410862791/pref ./mmseq_db/tmpDir/185048956948561117/linclust/3316508196410862791/pref_filter1 -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to pref_filter1: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 1ms\n",
      "filterdb ./mmseq_db/tmpDir/185048956948561117/linclust/3316508196410862791/pref_filter1 ./mmseq_db/tmpDir/185048956948561117/linclust/3316508196410862791/pref_filter2 --filter-file ./mmseq_db/tmpDir/185048956948561117/linclust/3316508196410862791/order_redundancy --threads 22 --compressed 0 -v 3 \n",
      "\n",
      "Filtering using file(s)\n",
      "[=================================================================] 1.74K 0s 5ms\n",
      "Time for merging to pref_filter2: 0h 0m 0s 1ms\n",
      "Time for processing: 0h 0m 0s 11ms\n",
      "rescorediagonal ./mmseq_db/tmpDir/185048956948561117/linclust/3316508196410862791/input_step_redundancy ./mmseq_db/tmpDir/185048956948561117/linclust/3316508196410862791/input_step_redundancy ./mmseq_db/tmpDir/185048956948561117/linclust/3316508196410862791/pref_filter2 ./mmseq_db/tmpDir/185048956948561117/linclust/3316508196410862791/pref_rescore2 --sub-mat nucl:nucleotide.out,aa:blosum62.out --rescore-mode 1 --wrapped-scoring 0 --filter-hits 1 -e 0.001 -c 0.8 -a 0 --cov-mode 1 --min-seq-id 0.3 --min-aln-len 0 --seq-id-mode 0 --add-self-matches 0 --sort-results 0 --db-load-mode 0 --threads 22 --compressed 0 -v 3 \n",
      "\n",
      "[=================================================================] 1.74K 0s 4ms\n",
      "Time for merging to pref_rescore2: 0h 0m 0s 1ms\n",
      "Time for processing: 0h 0m 0s 16ms\n",
      "align ./mmseq_db/tmpDir/185048956948561117/linclust/3316508196410862791/input_step_redundancy ./mmseq_db/tmpDir/185048956948561117/linclust/3316508196410862791/input_step_redundancy ./mmseq_db/tmpDir/185048956948561117/linclust/3316508196410862791/pref_rescore2 ./mmseq_db/tmpDir/185048956948561117/linclust/3316508196410862791/aln --sub-mat nucl:nucleotide.out,aa:blosum62.out -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.3 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 1 --max-seq-len 65535 --comp-bias-corr 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca 1 --pcb 1.5 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --gap-open nucl:5,aa:11 --gap-extend nucl:2,aa:1 --zdrop 40 --threads 22 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 1737 type: Aminoacid\n",
      "Target database size: 1737 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 1.74K 0s 54ms\n",
      "Time for merging to aln: 0h 0m 0s 1ms\n",
      "2247 alignments calculated\n",
      "2022 sequence pairs passed the thresholds (0.899866 of overall calculated)\n",
      "1.164076 hits per query sequence\n",
      "Time for processing: 0h 0m 0s 62ms\n",
      "clust ./mmseq_db/tmpDir/185048956948561117/linclust/3316508196410862791/input_step_redundancy ./mmseq_db/tmpDir/185048956948561117/linclust/3316508196410862791/aln ./mmseq_db/tmpDir/185048956948561117/linclust/3316508196410862791/clust --cluster-mode 3 --max-iterations 1000 --similarity-type 2 --threads 22 --compressed 0 -v 3 \n",
      "\n",
      "Clustering mode: Greedy Low Mem\n",
      "Total time: 0h 0m 0s 25ms\n",
      "\n",
      "Size of the sequence database: 1737\n",
      "Size of the alignment database: 1737\n",
      "Number of clusters: 1549\n",
      "\n",
      "Writing results 0h 0m 0s 0ms\n",
      "Time for merging to clust: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 26ms\n",
      "mergeclusters ./mmseq_db/inputDB ./mmseq_db/tmpDir/185048956948561117/clu_redundancy ./mmseq_db/tmpDir/185048956948561117/linclust/3316508196410862791/pre_clust ./mmseq_db/tmpDir/185048956948561117/linclust/3316508196410862791/clust --threads 22 --compressed 0 -v 3 \n",
      "\n",
      "Clustering step 1\n",
      "[=================================================================] 1.74K 0s 13ms\n",
      "Clustering step 2\n",
      "[=================================================================] 1.55K 0s 35ms\n",
      "Write merged clustering\n",
      "[=================================================================] 11.20K 0s 45ms\n",
      "Time for merging to clu_redundancy: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 57ms\n",
      "createsubdb ./mmseq_db/tmpDir/185048956948561117/clu_redundancy ./mmseq_db/inputDB ./mmseq_db/tmpDir/185048956948561117/input_step_redundancy -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step_redundancy: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 1ms\n",
      "prefilter ./mmseq_db/tmpDir/185048956948561117/input_step_redundancy ./mmseq_db/tmpDir/185048956948561117/input_step_redundancy ./mmseq_db/tmpDir/185048956948561117/pref_step0 --sub-mat nucl:nucleotide.out,aa:blosum62.out --seed-sub-mat nucl:nucleotide.out,aa:VTML80.out -s 1 -k 0 --k-score 2147483647 --alph-size nucl:5,aa:21 --max-seq-len 65535 --max-seqs 20 --split 0 --split-mode 2 --split-memory-limit 0 -c 0.8 --cov-mode 1 --comp-bias-corr 0 --diag-score 0 --exact-kmer-matching 0 --mask 1 --mask-lower-case 0 --min-ungapped-score 0 --add-self-matches 0 --spaced-kmer-mode 1 --db-load-mode 0 --pca 1 --pcb 1.5 --threads 22 --compressed 0 -v 3 \n",
      "\n",
      "Query database size: 1549 type: Aminoacid\n",
      "Estimated memory consumption: 983M\n",
      "Target database size: 1549 type: Aminoacid\n",
      "Index table k-mer threshold: 154 at k-mer size 6 \n",
      "Index table: counting k-mers\n",
      "[=================================================================] 1.55K 0s 20ms\n",
      "Index table: Masked residues: 6975\n",
      "Index table: fill\n",
      "[=================================================================] 1.55K 0s 11ms\n",
      "Index statistics\n",
      "Entries:          282028\n",
      "DB size:          489 MB\n",
      "Avg k-mer size:   0.004407\n",
      "Top 10 k-mers\n",
      "    SGVRHM\t32\n",
      "    RALNFQ\t23\n",
      "    RWNTRA\t22\n",
      "    KLPYLD\t22\n",
      "    ITFTWD\t22\n",
      "    RNKKYD\t22\n",
      "    LIMFGE\t22\n",
      "    QMEGGF\t22\n",
      "    NRKVTF\t22\n",
      "    FQGTAG\t22\n",
      "Time for index table init: 0h 0m 0s 262ms\n",
      "Process prefiltering step 1 of 1\n",
      "\n",
      "k-mer similarity threshold: 154\n",
      "Starting prefiltering scores calculation (step 1 of 1)\n",
      "Query db start 1 to 1549\n",
      "Target db start 1 to 1549\n",
      "[=================================================================] 1.55K 0s 17ms\n",
      "\n",
      "2.586344 k-mers per position\n",
      "230 DB matches per sequence\n",
      "0 overflows\n",
      "0 queries produce too many hits (truncated result)\n",
      "2 sequences passed prefiltering per query sequence\n",
      "1 median result list length\n",
      "0 sequences with 0 size result lists\n",
      "Time for merging to pref_step0: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 528ms\n",
      "align ./mmseq_db/tmpDir/185048956948561117/input_step_redundancy ./mmseq_db/tmpDir/185048956948561117/input_step_redundancy ./mmseq_db/tmpDir/185048956948561117/pref_step0 ./mmseq_db/tmpDir/185048956948561117/aln_step0 --sub-mat nucl:nucleotide.out,aa:blosum62.out -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.3 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 1 --max-seq-len 65535 --comp-bias-corr 0 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca 1 --pcb 1.5 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --gap-open nucl:5,aa:11 --gap-extend nucl:2,aa:1 --zdrop 40 --threads 22 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 1549 type: Aminoacid\n",
      "Target database size: 1549 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 1.55K 0s 74ms\n",
      "Time for merging to aln_step0: 0h 0m 0s 2ms\n",
      "3854 alignments calculated\n",
      "2328 sequence pairs passed the thresholds (0.604048 of overall calculated)\n",
      "1.502905 hits per query sequence\n",
      "Time for processing: 0h 0m 0s 81ms\n",
      "clust ./mmseq_db/tmpDir/185048956948561117/input_step_redundancy ./mmseq_db/tmpDir/185048956948561117/aln_step0 ./mmseq_db/tmpDir/185048956948561117/clu_step0 --cluster-mode 3 --max-iterations 1000 --similarity-type 2 --threads 22 --compressed 0 -v 3 \n",
      "\n",
      "Clustering mode: Greedy Low Mem\n",
      "Total time: 0h 0m 0s 16ms\n",
      "\n",
      "Size of the sequence database: 1549\n",
      "Size of the alignment database: 1549\n",
      "Number of clusters: 1321\n",
      "\n",
      "Writing results 0h 0m 0s 0ms\n",
      "Time for merging to clu_step0: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 18ms\n",
      "createsubdb ./mmseq_db/tmpDir/185048956948561117/clu_step0 ./mmseq_db/tmpDir/185048956948561117/input_step_redundancy ./mmseq_db/tmpDir/185048956948561117/input_step1 -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step1: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "prefilter ./mmseq_db/tmpDir/185048956948561117/input_step1 ./mmseq_db/tmpDir/185048956948561117/input_step1 ./mmseq_db/tmpDir/185048956948561117/pref_step1 --sub-mat nucl:nucleotide.out,aa:blosum62.out --seed-sub-mat nucl:nucleotide.out,aa:VTML80.out -s 3 -k 0 --k-score 2147483647 --alph-size nucl:5,aa:21 --max-seq-len 65535 --max-seqs 20 --split 0 --split-mode 2 --split-memory-limit 0 -c 0.8 --cov-mode 1 --comp-bias-corr 1 --diag-score 1 --exact-kmer-matching 0 --mask 1 --mask-lower-case 0 --min-ungapped-score 15 --add-self-matches 0 --spaced-kmer-mode 1 --db-load-mode 0 --pca 1 --pcb 1.5 --threads 22 --compressed 0 -v 3 \n",
      "\n",
      "Query database size: 1321 type: Aminoacid\n",
      "Estimated memory consumption: 982M\n",
      "Target database size: 1321 type: Aminoacid\n",
      "Index table k-mer threshold: 136 at k-mer size 6 \n",
      "Index table: counting k-mers\n",
      "[=================================================================] 1.32K 0s 22ms\n",
      "Index table: Masked residues: 6273\n",
      "Index table: fill\n",
      "[=================================================================] 1.32K 0s 19ms\n",
      "Index statistics\n",
      "Entries:          459963\n",
      "DB size:          490 MB\n",
      "Avg k-mer size:   0.007187\n",
      "Top 10 k-mers\n",
      "    SSLPSH\t31\n",
      "    SGVRHM\t30\n",
      "    HPFKGS\t12\n",
      "    DATMAA\t11\n",
      "    TKRADA\t11\n",
      "    ADAQDA\t11\n",
      "    DIVQDA\t11\n",
      "    NDLVKA\t11\n",
      "    GQDALA\t11\n",
      "    LKIKNA\t11\n",
      "Time for index table init: 0h 0m 0s 269ms\n",
      "Process prefiltering step 1 of 1\n",
      "\n",
      "k-mer similarity threshold: 136\n",
      "Starting prefiltering scores calculation (step 1 of 1)\n",
      "Query db start 1 to 1321\n",
      "Target db start 1 to 1321\n",
      "[=================================================================] 1.32K 0s 47ms\n",
      "\n",
      "23.724919 k-mers per position\n",
      "400 DB matches per sequence\n",
      "0 overflows\n",
      "0 queries produce too many hits (truncated result)\n",
      "3 sequences passed prefiltering per query sequence\n",
      "1 median result list length\n",
      "0 sequences with 0 size result lists\n",
      "Time for merging to pref_step1: 0h 0m 0s 1ms\n",
      "Time for processing: 0h 0m 0s 574ms\n",
      "align ./mmseq_db/tmpDir/185048956948561117/input_step1 ./mmseq_db/tmpDir/185048956948561117/input_step1 ./mmseq_db/tmpDir/185048956948561117/pref_step1 ./mmseq_db/tmpDir/185048956948561117/aln_step1 --sub-mat nucl:nucleotide.out,aa:blosum62.out -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.3 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 1 --max-seq-len 65535 --comp-bias-corr 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca 1 --pcb 1.5 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --gap-open nucl:5,aa:11 --gap-extend nucl:2,aa:1 --zdrop 40 --threads 22 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 1321 type: Aminoacid\n",
      "Target database size: 1321 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 1.32K 0s 51ms\n",
      "Time for merging to aln_step1: 0h 0m 0s 1ms\n",
      "3226 alignments calculated\n",
      "1532 sequence pairs passed the thresholds (0.474892 of overall calculated)\n",
      "1.159727 hits per query sequence\n",
      "Time for processing: 0h 0m 0s 57ms\n",
      "clust ./mmseq_db/tmpDir/185048956948561117/input_step1 ./mmseq_db/tmpDir/185048956948561117/aln_step1 ./mmseq_db/tmpDir/185048956948561117/clu_step1 --cluster-mode 3 --max-iterations 1000 --similarity-type 2 --threads 22 --compressed 0 -v 3 \n",
      "\n",
      "Clustering mode: Greedy Low Mem\n",
      "Total time: 0h 0m 0s 32ms\n",
      "\n",
      "Size of the sequence database: 1321\n",
      "Size of the alignment database: 1321\n",
      "Number of clusters: 1224\n",
      "\n",
      "Writing results 0h 0m 0s 0ms\n",
      "Time for merging to clu_step1: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 33ms\n",
      "createsubdb ./mmseq_db/tmpDir/185048956948561117/clu_step1 ./mmseq_db/tmpDir/185048956948561117/input_step1 ./mmseq_db/tmpDir/185048956948561117/input_step2 -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step2: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "prefilter ./mmseq_db/tmpDir/185048956948561117/input_step2 ./mmseq_db/tmpDir/185048956948561117/input_step2 ./mmseq_db/tmpDir/185048956948561117/pref_step2 --sub-mat nucl:nucleotide.out,aa:blosum62.out --seed-sub-mat nucl:nucleotide.out,aa:VTML80.out -s 5 -k 0 --k-score 2147483647 --alph-size nucl:5,aa:21 --max-seq-len 65535 --max-seqs 20 --split 0 --split-mode 2 --split-memory-limit 0 -c 0.8 --cov-mode 1 --comp-bias-corr 1 --diag-score 1 --exact-kmer-matching 0 --mask 1 --mask-lower-case 0 --min-ungapped-score 15 --add-self-matches 0 --spaced-kmer-mode 1 --db-load-mode 0 --pca 1 --pcb 1.5 --threads 22 --compressed 0 -v 3 \n",
      "\n",
      "Query database size: 1224 type: Aminoacid\n",
      "Estimated memory consumption: 982M\n",
      "Target database size: 1224 type: Aminoacid\n",
      "Index table k-mer threshold: 118 at k-mer size 6 \n",
      "Index table: counting k-mers\n",
      "[=================================================================] 1.22K 0s 22ms\n",
      "Index table: Masked residues: 6007\n",
      "Index table: fill\n",
      "[=================================================================] 1.22K 0s 12ms\n",
      "Index statistics\n",
      "Entries:          452770\n",
      "DB size:          490 MB\n",
      "Avg k-mer size:   0.007075\n",
      "Top 10 k-mers\n",
      "    SSLPSH\t29\n",
      "    SGVRHM\t28\n",
      "    HPFKGS\t12\n",
      "    FEGGGG\t9\n",
      "    AASLAI\t9\n",
      "    LDVRLI\t9\n",
      "    ELTVIL\t9\n",
      "    KRDANL\t9\n",
      "    SGDGNL\t9\n",
      "    SSVLEN\t9\n",
      "Time for index table init: 0h 0m 0s 283ms\n",
      "Process prefiltering step 1 of 1\n",
      "\n",
      "k-mer similarity threshold: 118\n",
      "Starting prefiltering scores calculation (step 1 of 1)\n",
      "Query db start 1 to 1224\n",
      "Target db start 1 to 1224\n",
      "[=================================================================] 1.22K 0s 203ms\n",
      "\n",
      "176.558966 k-mers per position\n",
      "813 DB matches per sequence\n",
      "0 overflows\n",
      "0 queries produce too many hits (truncated result)\n",
      "5 sequences passed prefiltering per query sequence\n",
      "4 median result list length\n",
      "0 sequences with 0 size result lists\n",
      "Time for merging to pref_step2: 0h 0m 0s 1ms\n",
      "Time for processing: 0h 0m 0s 752ms\n",
      "align ./mmseq_db/tmpDir/185048956948561117/input_step2 ./mmseq_db/tmpDir/185048956948561117/input_step2 ./mmseq_db/tmpDir/185048956948561117/pref_step2 ./mmseq_db/tmpDir/185048956948561117/aln_step2 --sub-mat nucl:nucleotide.out,aa:blosum62.out -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0.3 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 1 --max-seq-len 65535 --comp-bias-corr 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca 1 --pcb 1.5 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --gap-open nucl:5,aa:11 --gap-extend nucl:2,aa:1 --zdrop 40 --threads 22 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 1224 type: Aminoacid\n",
      "Target database size: 1224 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 1.22K 0s 71ms\n",
      "Time for merging to aln_step2: 0h 0m 0s 1ms\n",
      "5008 alignments calculated\n",
      "1298 sequence pairs passed the thresholds (0.259185 of overall calculated)\n",
      "1.060457 hits per query sequence\n",
      "Time for processing: 0h 0m 0s 81ms\n",
      "clust ./mmseq_db/tmpDir/185048956948561117/input_step2 ./mmseq_db/tmpDir/185048956948561117/aln_step2 ./mmseq_db/tmpDir/185048956948561117/clu_step2 --cluster-mode 3 --max-iterations 1000 --similarity-type 2 --threads 22 --compressed 0 -v 3 \n",
      "\n",
      "Clustering mode: Greedy Low Mem\n",
      "Total time: 0h 0m 0s 20ms\n",
      "\n",
      "Size of the sequence database: 1224\n",
      "Size of the alignment database: 1224\n",
      "Number of clusters: 1185\n",
      "\n",
      "Writing results 0h 0m 0s 0ms\n",
      "Time for merging to clu_step2: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 22ms\n",
      "mergeclusters ./mmseq_db/inputDB ./mmseq_db/clusterDB ./mmseq_db/tmpDir/185048956948561117/clu_redundancy ./mmseq_db/tmpDir/185048956948561117/clu_step0 ./mmseq_db/tmpDir/185048956948561117/clu_step1 ./mmseq_db/tmpDir/185048956948561117/clu_step2 --threads 22 --compressed 0 -v 3 \n",
      "\n",
      "Clustering step 1\n",
      "[=================================================================] 1.55K 0s 20ms\n",
      "Clustering step 2\n",
      "[=================================================================] 1.32K 0s 44ms\n",
      "Clustering step 3\n",
      "[=================================================================] 1.22K 0s 65ms\n",
      "Clustering step 4\n",
      "[=================================================================] 1.18K 0s 87ms\n",
      "Write merged clustering\n",
      "[=================================================================] 11.20K 0s 109ms\n",
      "Time for merging to clusterDB: 0h 0m 0s 2ms\n",
      "Time for processing: 0h 0m 0s 119ms\n",
      "result2repseq ./mmseq_db/inputDB ./mmseq_db/clusterDB ./mmseq_db/repDB \n",
      "\n",
      "MMseqs Version:\t13.45111\n",
      "Preload mode\t0\n",
      "Compressed  \t0\n",
      "Threads     \t22\n",
      "Verbosity   \t3\n",
      "\n",
      "[=================================================================] 1.18K 0s 6ms\n",
      "Time for merging to repDB: 0h 0m 0s 1ms\n",
      "Time for processing: 0h 0m 0s 14ms\n",
      "convert2fasta ./mmseq_db/repDB ./fasta/high_resolution_2dot8_membrane_protein_sequences_after_2000_representatives.fasta \n",
      "\n",
      "MMseqs Version:\t13.45111\n",
      "Use header DB\tfalse\n",
      "Verbosity    \t3\n",
      "\n",
      "Start writing file to ./fasta/high_resolution_2dot8_membrane_protein_sequences_after_2000_representatives.fasta\n",
      "Time for processing: 0h 0m 0s 1ms\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fasta_path = './fasta/high_resolution_2dot8_membrane_protein_sequences_after_2000'\n",
    "# Ensure tmpDir exists\n",
    "os.makedirs('./mmseq_db/tmpDir', exist_ok=True)\n",
    "\n",
    "# Step 1: create sequence DB\n",
    "os.system(f'mmseqs createdb {fasta_path}.fasta ./mmseq_db/inputDB')\n",
    "\n",
    "# Step 2: cluster sequences\n",
    "os.system('mmseqs cluster ./mmseq_db/inputDB ./mmseq_db/clusterDB ./mmseq_db/tmpDir --min-seq-id 0.3 -c 0.8 --cov-mode 1')\n",
    "\n",
    "# Step 3: extract representative sequences\n",
    "os.system('mmseqs result2repseq ./mmseq_db/inputDB ./mmseq_db/clusterDB ./mmseq_db/repDB')\n",
    "\n",
    "# Step 4: convert the repDB to FASTA\n",
    "os.system(f'mmseqs convert2fasta ./mmseq_db/repDB {fasta_path}_representatives.fasta')\n",
    "\n",
    "# Clear mmseq_db dir\n",
    "os.system('rm mmseq_db/* -rf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "7f87a559-b4e2-4a99-8d4e-3a3bdbc3160e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1185 1185\n"
     ]
    }
   ],
   "source": [
    "with open(f'{fasta_path}_representatives.fasta', 'r') as f:\n",
    "    pdb_chain = []\n",
    "    for line in f.readlines():\n",
    "        if line[0]=='>':\n",
    "            pdb_chain.append(line[1:])\n",
    "pdb_chain_arr = np.array([i.split('_') for i in pdb_chain])\n",
    "print(len(pdb_chain_arr), len(np.unique(pdb_chain_arr[:,0])))\n",
    "#np.save('mem_prot_rep_2dot8.npy', pdb_chain_arr)\n",
    "np.save('prot_rep_2dot8.npy', pdb_chain_arr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "a2d62ae0-6cdb-4415-b219-edcbd9535f08",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "17704 16683\n"
     ]
    }
   ],
   "source": [
    "with open('./fasta/high_resolution_2_protein_sequences_after_2000_representatives.fasta', 'r') as f:\n",
    "    pdb_chain = []\n",
    "    for line in f.readlines():\n",
    "        if line[0]=='>':\n",
    "            pdb_chain.append(line[1:])\n",
    "pdb_chain_arr = np.array([i.split('_') for i in pdb_chain])\n",
    "print(len(pdb_chain_arr), len(np.unique(pdb_chain_arr[:,0])))\n",
    "np.save('prot_rep.npy', pdb_chain_arr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "394b0ffa-aad3-4cc0-a60a-6897b42950ff",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.23"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
